---
title: "Topographic, single-cell gene expression profiling of Periaqueductal Gray neurons"
subtitle: "Part VIII: data visualisation"
author:
  - name: "Oriol Pavon Arocas, Sarah F. Olesen, and Tiago Branco"
    affiliation: "Sainsbury Wellcome Centre for Neural Circuits and Behaviour, University College London, UK"
    email: "oriol.pavon.16@ucl.ac.uk"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook:
    highlight: pygments
    number_sections: FALSE
    theme: lumen
    toc: TRUE
    toc_float: TRUE

#output: rmdformats::readthedown:
  #highlight: pygments
---

***

This is a pipeline to analyse single-cell RNA sequencing data from Periaqueductal Gray neurons (1) isolated from acute midbrain slices of transgenic mice using visually guided aspiration via patch pipettes and (2) processed using SMART-seq2 (Picelli et al., Nature Protocols 2014).

This pipeline has been generated after attending the [EMBL-EBI RNA-Sequence Analysis Course](https://www.ebi.ac.uk/training/events/2019/rna-sequence-analysis) and [attending](https://training.csx.cam.ac.uk/bioinformatics/event/2823386) and following the online course on [Analysis of single cell RNA-seq data](https://github.com/hemberg-lab/scRNA.seq.course) by the [Hemberg Lab](https://www.sanger.ac.uk/science/groups/hemberg-group). Many other resources have been used, including the [Orchestrating Single-Cell Analysis with Bioconductor book](https://bioconductor.org/books/release/OSCA/) by Robert Amezquita, Aaron Lun, Stephanie Hicks, and Raphael Gottardo, the [simpleSingleCell workflow in Bioconductor](https://bioconductor.org/packages/3.9/workflows/html/simpleSingleCell.html) maintained by Aaron Lun, the [rnaseqGene workflow](https://bioconductor.org/packages/3.10/workflows/html/rnaseqGene.html) maintained by Michael Love, the [RNAseq123 workflow](https://bioconductor.org/packages/3.10/workflows/html/RNAseq123.html) maintained by Matthew Ritchie, and the [EGSEA123 workflow](https://bioconductor.org/packages/3.10/workflows/html/EGSEA123.html) maintained by Matthew Ritchie.

Other key resources include [Bioconductor](http://www.bioconductor.org/) (Huber et al., Nature Methods 2015), [scRNA-tools](https://www.scrna-tools.org/), `scater` (McCarthy et al., Bioinformatics 2017), `scran` (Lun et al., F1000Res 2016), `SC3` (Kiselev et al., Nature Methods 2017), `Seurat` (Butler et al., Nature Biotechnology 2018), `clusterExperiment` (Risso et al., PLOS Computational Biology 2018), `limma` (Ritchie et al., Nucleic Acids Research 2015), `DESeq2` (Love et al., Genome Biology 2014), `edgeR` (Robinson et al., Bioinformatics 2010), `MAST` (Finak, McDavid, Yajima et al., Genome Biology 2015), `iSEE` (Rue-Albrecht & Marini et al., F1000Research 2018), `t-SNE` (van der Maaten & Hinton, Journal of Machine Learning Research 2008), `UMAP` (McInnes et al., arXiv 2018), and the [Mathematical Statistics and Machine Learning for Life Sciences](https://towardsdatascience.com/tagged/stats-ml-life-sciences) column by Nikolay Oskolkov.

***

## Step 8.1 | UMAP projection with clustering results
```{r}
# Set the directory where your data and scripts are:
setwd("D:/Dropbox (UCL - SWC)/Project_transcriptomics/analysis/PAG_scRNAseq_analysis")

# Set the path to save figures from this Part:
path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part8_datavis/"
date <- Sys.Date()
date <- gsub("-", "_", date)

# Load packages:
library(tidyverse)
library(SingleCellExperiment)
library(scater)
library(scran)
library(ggplot2)
library(pheatmap)
library(DESeq2)
library(edgeR)
library(limma)
library(monocle)
library(MAST)
library(ROCR)
library(patchwork)
library(extrafont)
```

To recapitulate the plots with the clustering results, we load the `PAG_sceset_qc_norm_filt_corr_clust` after normalisation, filtering, batch correction, and clustering. We should thus have a `corrected` slot in `assays`:
```{r}
# If starting from stored results, load saved filtered dataset from previous Step:
options(stringsAsFactors = FALSE)

PAG_sceset_qc_norm_filt_corr_clust <- readRDS("PAG_sceset_qc_norm_filt_corr_clust.rds") # Contains filtered cells and genes, log-normalised, filtered, corrected, and clustered data
assayNames(PAG_sceset_qc_norm_filt_corr_clust)
reducedDimNames(PAG_sceset_qc_norm_filt_corr_clust)
```

To remind ourselves about which clustering results we have available, we can look at the names of the `colData` slot:
```{r}
names(colData(PAG_sceset_qc_norm_filt_corr_clust)[grep("cluster", names(colData(PAG_sceset_qc_norm_filt_corr_clust)))])
```

We can quickly replot the results of our chosen clustering solutions:
```{r}
# Set theme parameters
theme_args_clustering <- theme(plot.title = element_text(size = 11, face = "bold"),
                               axis.title = element_text(size = 11, face = "plain"),
                               axis.text = element_text(size = 10, face = "plain"), 
                               legend.text = element_text(size = 9, face = "plain"), 
                               strip.text = element_text(size = 10, face = "plain"))


### Using UMAP (CV2) - 15_neighbors_0.01_min_dist ###
UMAP_cv2_15_001_PAGsubdivisions <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                  dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                  colour_by = "PAG.area", shape_by = "cell.type",
                                                  point_alpha = 0.6, point_size = 2, theme_size = 11
                                                  ) + labs(title = "PAG subdivisions", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

### Graph-based clustering results ###
# Using HVG from CV2 with rank-based weights
UMAP_cv2_15_001_RankWalktrap_k7 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                  dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                  colour_by = "SNN_clusters_cv2_rank_k7", shape_by = "cell.type", text_by = "SNN_clusters_cv2_rank_k7",
                                                  point_alpha = 0.6, point_size = 2, theme_size = 11
                                                  ) + labs(title = "Rank-Walktrap (k = 7)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_RankWalktrap_k12 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                  dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                  colour_by = "SNN_clusters_cv2_rank_k12", shape_by = "cell.type", text_by = "SNN_clusters_cv2_rank_k12",
                                                  point_alpha = 0.6, point_size = 2, theme_size = 11
                                                  ) + labs(title = "Rank-Walktrap (k = 12)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

# Using HVG from CV2 with jaccard-based weights
UMAP_cv2_15_001_JaccardLouvain_k5 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k5", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k5",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 5)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_JaccardLouvain_k8 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k8", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k8",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 8)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_JaccardLouvain_k9 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k9", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k9",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 9)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_JaccardLouvain_k13 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k13", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k13",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 13)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

### K-means clustering results ###
### Hierarchical clustering results ###
# The clusters obtained with both the K-means and the hierarchical clustering approach do not match well with the UMAP representation of the data. We exclude them from here but can be seen in Part 6 and readded if necessary.

### SC3 clustering results ###
UMAP_cv2_15_001_SC3_4 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "sc3_4_clusters", shape_by = "cell.type", text_by = "sc3_4_clusters",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering



# Prepare and save composed plots
library(patchwork)

UMAP_SNN_RankWalktrap_k7 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_RankWalktrap_k7) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_RankWalktrap_k7

UMAP_SNN_RankWalktrap_k12 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_RankWalktrap_k12) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_RankWalktrap_k12

UMAP_SNN_JaccardLouvain_k5 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k5) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k5

UMAP_SNN_JaccardLouvain_k8 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k8) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k8

UMAP_SNN_JaccardLouvain_k9 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k9) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k9

UMAP_SNN_JaccardLouvain_k13 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k13) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k13

UMAP_SC3_k4 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_SC3_4) +
  plot_annotation(title = "SC3 clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SC3_k4


# # We could save the plots as follows, although we have already done so in earlier parts of the pipeline.
# ggsave(filename = str_c(date, "_UMAP_SNN_rank_k7_clusters.pdf"),
#        plot = UMAP_SNN_RankWalktrap_k7,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_rank_k12_clusters.pdf"),
#        plot = UMAP_SNN_RankWalktrap_k12,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k5_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k5,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k8_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k8,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k9_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k9,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k13_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k13,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_sc3_k4_clusters.pdf"),
#        plot = UMAP_SC3_k4,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
```

## Step 8.2 | Visualising results from differential expression analysis
```{r}
library(pheatmap)
library(RColorBrewer)
library(viridis)
best_set_wilcox_up_all_SC3_cluster_4 <- marker_set_SC3_k4_4_wilcox_all_up[marker_set_SC3_k4_4_wilcox_all_up$FDR<0.05,]
rownames(best_set_wilcox_up_all_SC3_cluster_4)
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = rownames(best_set_wilcox_up_all_SC3_cluster_4)[1:15], # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters), 
            # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(-5, 5), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix.
            # Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("sc3_4_clusters", "cell.type", "PAG.area"), # how the columns should be annotated with colours.
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = "Marker genes for SC3 Cluster 4 (wilcoxon | all | up)")
```

```{r}
best_set_wilcox_up_all_SC3_cluster_4 <- marker_set_SC3_k4_4_wilcox_all_up[marker_set_SC3_k4_4_wilcox_all_up$FDR<0.05,]
#best_set_wilcox_up_all_SC3_cluster_4
AUCs_wilcox_up <- as.matrix(best_set_wilcox_up_all_SC3_cluster_4[,-(1:2)])
#AUCs_wilcox_up
colnames(AUCs_wilcox_up) <- sub("AUC.", "cluster_", colnames(AUCs_wilcox_up)) # Remove "AUC." from the colnames, to leave the cluster number alone
#AUCs_wilcox_up

# Plot the AUCs - will plot one column for each comparison, and the top marker genes for the selected group should have a high value for each comparison
pheatmap(AUCs_wilcox_up, breaks = seq(0, 1, length.out = 21), color = viridis::viridis(21),
         main = "Heatmap for Cluster 4 AUCs (wilcoxon | all | up)")

# Alternative plot - will plot expression of the marker genes
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = rownames(best_set_wilcox_up_all_SC3_cluster_4)[1:15], # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters), 
            # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(-5, 5), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix.
            # Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("sc3_4_clusters", "cell.type", "PAG.area"), # how the columns should be annotated with colours.
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = "Marker genes for SC3 Cluster 4 (wilcoxon | all | up)")
```

```{r}
# Colours PAG
colours_celltype_PAGarea <- c(`dmpag_VGAT` = "red", `dlpag_VGAT` = "coral2", `lpag_VGAT` = "salmon", `vlpag_VGAT` = "darksalmon",
                              `dmpag_VGluT2` = "cornflowerblue", `dlpag_VGluT2` = "cyan3", `lpag_VGluT2` = "aquamarine3", `vlpag_VGluT2` = "cadetblue3")

colours_celltype = c(rgb(255, 119, 124, 255/2, maxColorValue = 255), rgb(0, 156, 181, 255/6, maxColorValue = 255))

# Heatmap palettes
heatmap_viridis = rev(viridis::magma(100))
heatmap_blue = colorRampPalette(c("white", "blue"))(100)
heatmap_VGAT <- colorRampPalette(c("white", "#FF8080"))(100)
heatmap_VGluT2 <- colorRampPalette(c("white", "#0F99B2"))(100)

# Clustering colours:
cluster_colours_10 <- c(`1` = "#729ECE", `2` = "#FF9E4A", `3` = "#67BF5C", `4` = "#ED665D", `5` = "#AD8BC9",
                        `6` = "#A8786E", `7` = "#ED97CA", `8` = "#A2A2A2", `9` = "#CDCC5D", `10` = "#6DCCDA")

cluster_colours_20 <- c(`1` = "#1F77B4", `2` = "#AEC7E8", `3` = "#FF7F0E", `4` = "#FFBB78", `5` = "#2CA02C",
                        `6` = "#98DF8A", `7` = "#D62728", `8` = "#FF9896", `9` = "#9467BD", `10` = "#C5B0D5",
                        `11` = "#8C564B", `12` = "#C49C94", `13` = "#E377C2", `14` = "#F7B6D2", `15` = "#7F7F7F", 
                        `16` = "#C7C7C7", `17` = "#BCBD22", `18` = "#DBDB8D", `19` = "#17BECF", `20` = "#9EDAE5")

cluster_colors <- scater:::.get_palette("tableau10medium") # hidden scater colours


# Scater palettes:
tableau20 = c("#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78", "#2CA02C",
              "#98DF8A", "#D62728", "#FF9896", "#9467BD", "#C5B0D5",
              "#8C564B", "#C49C94", "#E377C2", "#F7B6D2", "#7F7F7F",
              "#C7C7C7", "#BCBD22", "#DBDB8D", "#17BECF", "#9EDAE5")

tableau10medium = c("#729ECE", "#FF9E4A", "#67BF5C", "#ED665D",
                    "#AD8BC9", "#A8786E", "#ED97CA", "#A2A2A2",
                    "#CDCC5D", "#6DCCDA")

colorblind10 = c("#006BA4", "#FF800E", "#ABABAB", "#595959",
                 "#5F9ED1", "#C85200", "#898989", "#A2C8EC",
                 "#FFBC79", "#CFCFCF")

trafficlight = c("#B10318", "#DBA13A", "#309343", "#D82526",
                 "#FFC156", "#69B764", "#F26C64", "#FFDD71",
                 "#9FCD99")

purplegray12 = c("#7B66D2", "#A699E8", "#DC5FBD", "#FFC0DA",
                 "#5F5A41", "#B4B19B", "#995688", "#D898BA",
                 "#AB6AD5", "#D098EE", "#8B7C6E", "#DBD4C5")

bluered12 = c("#2C69B0", "#B5C8E2", "#F02720", "#FFB6B0", "#AC613C",
              "#E9C39B", "#6BA3D6", "#B5DFFD", "#AC8763", "#DDC9B4",
              "#BD0A36", "#F4737A")

greenorange12 = c("#32A251", "#ACD98D", "#FF7F0F", "#FFB977",
                  "#3CB7CC", "#98D9E4", "#B85A0D", "#FFD94A",
                  "#39737C", "#86B4A9", "#82853B", "#CCC94D")

cyclic = c("#1F83B4", "#1696AC", "#18A188", "#29A03C", "#54A338",
           "#82A93F", "#ADB828", "#D8BD35", "#FFBD4C", "#FFB022",
           "#FF9C0E", "#FF810E", "#E75727", "#D23E4E", "#C94D8C",
           "#C04AA7", "#B446B3", "#9658B1", "#8061B4", "#6F63BB")
```


```{r}
# Annotating the heatmap (from https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/SingleCell/practical/crukBioinfoSummerSchoolJuly2018_scRnaSeqCellPopId_practical_run.html#192_detecting_genes_differentially_expressed_between_clusters)

tmpData <- logcounts(PAG_sceset_qc_norm_filt_corr_clust_de)[rownames(best_set_wilcox_up_all_SC3_cluster_4)[1:15],]
dim(tmpData)

# columns annotation with cell name:
mat_col <- data.frame(cell.type = PAG_sceset_qc_norm_filt_corr_clust_de$cell.type,
                      PAG.subdivision = PAG_sceset_qc_norm_filt_corr_clust_de$PAG.area,
                      cluster.ID = PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters)
rownames(mat_col) <- colnames(PAG_sceset_qc_norm_filt_corr_clust_de)
dim(mat_col)

# Prepare colours for clusters:
colourCount = length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters))
mat_colors <- list(cluster.ID = tableau10medium[1:length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters))], 
                   cell.type = c("#FF8080", "#0F99B2"),
                   PAG.subdivision = tableau10medium[1:length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$PAG.area))])
mat_colors

names(mat_colors$cluster.ID) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters))
names(mat_colors$cell.type) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))
names(mat_colors$PAG.subdivision) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$PAG.area))
mat_colors

library(pheatmap)
library(RColorBrewer)
library(viridis)
pheatmap(tmpData[,order(mat_col$cluster.ID)],
         border_color = NA,
         scale = "none",
         color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name =         "RdYlBu")))(100)
         show_rownames = TRUE, show_colnames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE,
         annotation_col = mat_col, 
         annotation_colors = mat_colors,
         main = "Marker genes for SC3 Cluster 4 (wilcoxon | all | up)")
```