---
title: "Topographic, single-cell gene expression profiling of Periaqueductal Gray neurons"
subtitle: "Violin and Heatmap plots"
author:
  - name: "Oriol Pavon Arocas, Sarah F. Olesen, and Tiago Branco"
    affiliation: "Sainsbury Wellcome Centre for Neural Circuits and Behaviour, University College London, UK"
    email: "oriol.pavon.16@ucl.ac.uk"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook:
    highlight: pygments
    number_sections: FALSE
    theme: lumen
    toc: TRUE
    toc_float: TRUE

#output: rmdformats::readthedown:
  #highlight: pygments
---

***
This is a pipeline to analyse single-cell RNA sequencing data from Periaqueductal Gray neurons (1) isolated from acute midbrain slices of transgenic mice using visually guided aspiration via patch pipettes and (2) processed using SMART-seq2 (Picelli et al. Nature Protocols 2014). 

This pipeline has been generated after attending the [EMBL-EBI RNA-Sequence Analysis Course](https://www.ebi.ac.uk/training/events/2019/rna-sequence-analysis) and [attending](https://training.csx.cam.ac.uk/bioinformatics/event/2823386) and following the online course on [Analysis of single cell RNA-seq data](https://github.com/hemberg-lab/scRNA.seq.course) by the [Hemberg Lab](https://www.sanger.ac.uk/science/groups/hemberg-group). Many other resources have been used, including the [Orchestrating Single-Cell Analysis with Bioconductor book](https://osca.bioconductor.org/) by Robert Amezquita and Stephanie Hicks, the [simpleSingleCell workflow in Bioconductor](https://bioconductor.org/packages/3.9/workflows/html/simpleSingleCell.html) maintained by Aaron Lun, the [rnaseqGene workflow](https://bioconductor.org/packages/release/workflows/html/rnaseqGene.html) maintained by Michael Love, the [RNAseq123 workflow](https://bioconductor.org/packages/release/workflows/html/RNAseq123.html) maintained by Matthew Ritchie, and the [EGSEA123 workflow](https://bioconductor.org/packages/release/workflows/html/EGSEA123.html) maintained by Matthew Ritchie.

Other key resources are Bioconductor (Huber et al., Nature Methods 2015), [scRNA-tools](https://www.scrna-tools.org/), `scater` (McCarty et al., Bioinformatics 2017), `scran` (Lun et al. F1000Res 2016), `SC3` (Kiselev et al., Nature Methods 2017), `Seurat` (Butler et al., Nature Biotechnology 2018), `clusterExperiment` (Risso et al., PLOS Computational Biology 2018), `limma` (Ritchie et al., Nucleic Acids Research 2015), `DESeq2` (Love et al., Genome Biology 2014), `iSEE` (Rue-Albrecht & Marini et al., F1000Research 2018), `t-SNE` (van der Maaten & Hinton, Journal of Machine Learning Research 2008).

***

# STEP X | Inspecting the data using violin and heatmaps plots
The following code aims at facilitating visualization by plotting the expression of different list of genes stored in the `SingleCellExperiment` object using heatmaps and violing plots.

Load `PAG_sceset_qc_norm` after clustering and differential expression analysis. We should thus have a `logcounts` slot in `assays`, as well as cluster identities and lists of HVG and DE genes. We should also still have our gene lists loaded into the `metadata` slot:
```{r}
# If starting from stored results, load saved filtered dataset from previous Step:
set.seed(1991)
options(stringsAsFactors = FALSE)
library(SingleCellExperiment)
library(scater)
library(scran)

PAG_sceset_qc_norm <- readRDS("PAG_sceset_qc_norm.rds") # Contains filtered cells and genes, normalized data, cluster IDs, and DE genes
assayNames(PAG_sceset_qc_norm)
PAG_sceset_qc_norm
names(metadata(PAG_sceset_qc_norm))
```

The following code chunks will follow the same structure. We first get the gene names we want to inspect from our previously loaded `PAG_scRNAseq_gene_lists`. We then filter any genes not present in our QCed dataset by comparing the gene names in the `metadata` list with the `rownames` of our expression matrix. We can next generate the violin or heatmap plots.

Some visualisation tricks:
* Use `PAG_ionchannels[1:20]` to select a subset of ion channels
* Alternate between `x = "PAG.areacollection", colour_by = "cell.type"` or `x = "cell.type", colour_by = "PAG.areacollection"`
* Use `ncol = 5` to speficy the number of columns the plot will be divided into

## Step X.1 | Ion Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_ionchannels <- metadata(PAG_sceset_qc_norm)$genes.ionchannels[(metadata(PAG_sceset_qc_norm)$genes.ionchannels %in% rownames(PAG_sceset_qc_norm))]
length(PAG_ionchannels)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_ionchannels[1:20],  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_ionchannels[1:20],  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_ionchannels[1:15],
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of ion channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_ionchannels[1:15], 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of ion channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_ionchannels[1:15], 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of ion channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_ionchannels[1:15],
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of ion channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_ionchannels[1:15], 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of ion channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_ionchannels[1:15],
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of ion channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Potassium Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_potassium <- metadata(PAG_sceset_qc_norm)$genes.icPotassium[(metadata(PAG_sceset_qc_norm)$genes.icPotassium %in% rownames(PAG_sceset_qc_norm))]
length(PAG_potassium)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_potassium,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_potassium,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_potassium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Potassium channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_potassium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Potassium channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_potassium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Potassium channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_potassium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Potassium channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_potassium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Potassium channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_potassium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Potassium channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Calcium Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_calcium <- metadata(PAG_sceset_qc_norm)$genes.icCalcium[(metadata(PAG_sceset_qc_norm)$genes.icCalcium %in% rownames(PAG_sceset_qc_norm))]
length(PAG_calcium)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_calcium,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_calcium,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_calcium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Calcium channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_calcium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Calcium channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_calcium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Calcium channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_calcium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Calcium channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_calcium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Calcium channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_calcium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Calcium channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Sodium Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_sodium <- metadata(PAG_sceset_qc_norm)$genes.icSodium[(metadata(PAG_sceset_qc_norm)$genes.icSodium %in% rownames(PAG_sceset_qc_norm))]
length(PAG_sodium)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_sodium,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_sodium,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_sodium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Sodium channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_sodium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Sodium channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_sodium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Sodium channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_sodium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Sodium channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_sodium, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Sodium channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_sodium,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Sodium channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Chloride Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_chloride <- metadata(PAG_sceset_qc_norm)$genes.icChloride[(metadata(PAG_sceset_qc_norm)$genes.icChloride %in% rownames(PAG_sceset_qc_norm))]
length(PAG_chloride)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_chloride,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_chloride,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_chloride,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Chloride channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_chloride, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Chloride channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_chloride, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Chloride channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_chloride,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Chloride channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_chloride, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Chloride channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_chloride,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Chloride channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Asic Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_asic <- metadata(PAG_sceset_qc_norm)$genes.icAsic[(metadata(PAG_sceset_qc_norm)$genes.icAsic %in% rownames(PAG_sceset_qc_norm))]
length(PAG_asic)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_asic,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_asic,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_asic,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Asic channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_asic, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Asic channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_asic, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Asic channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_asic,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Asic channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_asic, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Asic channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_asic,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Asic channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Cyclic Nucleotide Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_cyclic_nucleotides <- metadata(PAG_sceset_qc_norm)$genes.icCyclicnucleotides[(metadata(PAG_sceset_qc_norm)$genes.icCyclicnucleotides %in% rownames(PAG_sceset_qc_norm))]
length(PAG_cyclic_nucleotides)
 
# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_cyclic_nucleotides,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_cyclic_nucleotides,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_cyclic_nucleotides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Cyclic Nucelotide channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_cyclic_nucleotides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Cyclic Nucelotide channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_cyclic_nucleotides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Cyclic Nucelotide channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_cyclic_nucleotides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Cyclic Nucelotide channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_cyclic_nucleotides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Cyclic Nucelotide channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_cyclic_nucleotides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Cyclic Nucelotide channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### TRP Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_trp <- metadata(PAG_sceset_qc_norm)$genes.icTRP[(metadata(PAG_sceset_qc_norm)$genes.icTRP %in% rownames(PAG_sceset_qc_norm))]
length(PAG_trp)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_trp,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_trp,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_trp,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of TRP channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_trp, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of TRP channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_trp, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of TRP channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_trp,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of TRP channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_trp, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of TRP channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_trp,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of TRP channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Connexins Channels
```{r} 
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_connexins <- metadata(PAG_sceset_qc_norm)$genes.icConnexins[(metadata(PAG_sceset_qc_norm)$genes.icConnexins %in% rownames(PAG_sceset_qc_norm))]
length(PAG_connexins)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_connexins,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_connexins,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_connexins,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Connexin channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_connexins, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Connexin channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_connexins, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Connexin channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_connexins,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Connexin channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_connexins, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Connexin channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_connexins,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Connexin channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Purinergic Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_purinergic <- metadata(PAG_sceset_qc_norm)$genes.icP2X[(metadata(PAG_sceset_qc_norm)$genes.icP2X %in% rownames(PAG_sceset_qc_norm))]
length(PAG_purinergic)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_purinergic,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_purinergic,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_purinergic,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Purinergic channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_purinergic, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Purinergic channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_purinergic, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Purinergic channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_purinergic,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Purinergic channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_purinergic, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Purinergic channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_purinergic,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Purinergic channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Aquaporins Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_aquaporins <- metadata(PAG_sceset_qc_norm)$genes.icAquaporins[(metadata(PAG_sceset_qc_norm)$genes.icAquaporins %in% rownames(PAG_sceset_qc_norm))]
length(PAG_aquaporins)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_aquaporins,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_aquaporins,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_aquaporins,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Aquaporin channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_aquaporins, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Aquaporin channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_aquaporins, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Aquaporin channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_aquaporins,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Aquaporin channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_aquaporins, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Aquaporin channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_aquaporins,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Aquaporin channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Other Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_otherchannels <- metadata(PAG_sceset_qc_norm)$genes.icOther[(metadata(PAG_sceset_qc_norm)$genes.icOther %in% rownames(PAG_sceset_qc_norm))]
length(PAG_otherchannels)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_otherchannels,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_otherchannels,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_otherchannels,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Other channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_otherchannels, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Other channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_otherchannels, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Other channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_otherchannels,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Other channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_otherchannels, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Other channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_otherchannels,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Other channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

## Step X.2 | Neurotransmitter Receptors and Transporters
### GABA
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_gaba <- metadata(PAG_sceset_qc_norm)$genes.GABA[(metadata(PAG_sceset_qc_norm)$genes.GABA %in% rownames(PAG_sceset_qc_norm))]
length(PAG_gaba)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_gaba,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_gaba,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_gaba,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of GABA in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_gaba, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of GABA in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_gaba, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of GABA in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_gaba,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of GABA in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_gaba, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of GABA in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_gaba,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of GABA in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Glutamate
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_glutamate <- metadata(PAG_sceset_qc_norm)$genes.Glutamate[(metadata(PAG_sceset_qc_norm)$genes.Glutamate %in% rownames(PAG_sceset_qc_norm))]
length(PAG_glutamate)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_glutamate,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_glutamate,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_glutamate,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Glutamate in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_glutamate, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Glutamate in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_glutamate, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Glutamate in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_glutamate,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Glutamate in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_glutamate, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Glutamate in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_glutamate,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Glutamate in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Acetylcholine
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_acetylcholine <- metadata(PAG_sceset_qc_norm)$genes.Acetylcholine[(metadata(PAG_sceset_qc_norm)$genes.Acetylcholine %in% rownames(PAG_sceset_qc_norm))]
length(PAG_acetylcholine)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_acetylcholine,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_acetylcholine,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_acetylcholine,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Acetylcholine in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_acetylcholine, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Acetylcholine in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_acetylcholine, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Acetylcholine in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_acetylcholine,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Acetylcholine in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_acetylcholine, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Acetylcholine in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_acetylcholine,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Acetylcholine in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Monoamines
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_monoamines <- metadata(PAG_sceset_qc_norm)$genes.Monoamines[(metadata(PAG_sceset_qc_norm)$genes.Monoamines %in% rownames(PAG_sceset_qc_norm))]
length(PAG_monoamines)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_monoamines,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_monoamines,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_monoamines,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Monoamines in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_monoamines, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Monoamines in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_monoamines, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Monoamines in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_monoamines,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Monoamines in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_monoamines, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Monoamines in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_monoamines,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Monoamines in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Transporters
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_transporters <- metadata(PAG_sceset_qc_norm)$genes.Transporters[(metadata(PAG_sceset_qc_norm)$genes.Transporters %in% rownames(PAG_sceset_qc_norm))]
length(PAG_transporters)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_transporters,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_transporters,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_transporters,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Transporters in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_transporters, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Transporters in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_transporters, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Transporters in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_transporters,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Transporters in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_transporters, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Transporters in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_transporters,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Transporters in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

## Step X.3 | Neuromodulators and Neuropeptides
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Neuromodulators_Peptides <- metadata(PAG_sceset_qc_norm)$genes.NeuromodulatorsPeptides[(metadata(PAG_sceset_qc_norm)$genes.NeuromodulatorsPeptides %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Neuromodulators_Peptides)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Neuromodulators_Peptides,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Neuromodulators_Peptides,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Neuromodulators_Peptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuromodulators and Neuropeptides in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Neuromodulators_Peptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuromodulators and Neuropeptides in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Neuromodulators_Peptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuromodulators and Neuropeptides in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Neuromodulators_Peptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuromodulators and Neuropeptides in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Neuromodulators_Peptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuromodulators and Neuropeptides in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Neuromodulators_Peptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuromodulators and Neuropeptides in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```
## Step X.4 | Gene Lists from Literature and Labs
### Cell type markers
(i) oligodendrocytes (mRNAs for myelin basic protein (Mbp) and UDP glyco- syltransferase-8 (Ugt8) involved in sphingolipid metabolism); 
(ii) astrocytes (genes Fabp7 and Ntsr2); 
(iii) ependymal cells (genes associated with motile cilia, including Enkur9 and the transcription factor Foxj1); 
(iv) microglia (mRNAs for macrophage lineage marker allograft inflammatory factor 1 (Aif1) and immunoglobulin E receptor (Fcer1g)); 
(v) endothelial cells (genes encoding fibronectin (Fn1) and organic anion transporter (Slco1a4)); 
(vi) vascular and smooth muscle lineage (mRNAs for the 2 subunit of actin (Acta2) and transgelin (Tagln)); 
and (vii) neurons (genes Ndrg4 and Stmn2).
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_celltypes <- metadata(PAG_sceset_qc_norm)$genes.CellTypeMarkers[(metadata(PAG_sceset_qc_norm)$genes.CellTypeMarkers %in% rownames(PAG_sceset_qc_norm))]
length(PAG_celltypes)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_celltypes,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_celltypes,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_celltypes,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Cell Type markers  in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_celltypes, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Cell Type markers  in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_celltypes, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Cell Type markers  in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_celltypes,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Cell Type markers  in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_celltypes, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Cell Type markers  in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_celltypes,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Cell Type markers in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### DE genes from QMUL
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_DEgenesQMUL <- metadata(PAG_sceset_qc_norm)$genes.QMULDE[(metadata(PAG_sceset_qc_norm)$genes.QMULDE %in% rownames(PAG_sceset_qc_norm))]
length(PAG_DEgenesQMUL)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_DEgenesQMUL,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_DEgenesQMUL,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_DEgenesQMUL,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of DE genes from QMUL in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_DEgenesQMUL, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of DE genes from QMUL in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_DEgenesQMUL, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of DE genes from QMUL in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_DEgenesQMUL,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of DE genes from QMUL in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_DEgenesQMUL, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of DE genes from QMUL in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_DEgenesQMUL,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of DE genes from QMUL in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### DE genes from Miguel Branco
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_DEgenesMB <- metadata(PAG_sceset_qc_norm)$genes.MBrancoDE[(metadata(PAG_sceset_qc_norm)$genes.MBrancoDE %in% rownames(PAG_sceset_qc_norm))]
length(PAG_DEgenesMB)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_DEgenesMB,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_DEgenesMB,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_DEgenesMB,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of DE genes from Miguel Branco in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_DEgenesMB, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of DE genes from Miguel Branco in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_DEgenesMB, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of DE genes from Miguel Branco in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_DEgenesMB,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of DE genes from Miguel Branco in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_DEgenesMB, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of DE genes from Miguel Branco in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_DEgenesMB,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of DE genes from Miguel Branco in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Dulac Lab Neuromodulators
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Dulac_neuromodulators <- metadata(PAG_sceset_qc_norm)$genes.DulacNeuromodulators[(metadata(PAG_sceset_qc_norm)$genes.DulacNeuromodulators %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Dulac_neuromodulators)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Dulac_neuromodulators,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Dulac_neuromodulators,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Dulac_neuromodulators,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuromodulators in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Dulac_neuromodulators, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuromodulators in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Dulac_neuromodulators, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuromodulators in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Dulac_neuromodulators,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuromodulators in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Dulac_neuromodulators, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuromodulators in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Dulac_neuromodulators,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuromodulators in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Dulac Lab Neuropeptides
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Dulac_neuropeptides <- metadata(PAG_sceset_qc_norm)$genes.DulacNeuropeptides[(metadata(PAG_sceset_qc_norm)$genes.DulacNeuropeptides %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Dulac_neuropeptides)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Dulac_neuropeptides,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Dulac_neuropeptides,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Dulac_neuropeptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuropeptides in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Dulac_neuropeptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuropeptides in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Dulac_neuropeptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,Dulac Lab Neuropeptides
            main = "Expression of Dulac Lab Neuropeptides in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Dulac_neuropeptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuropeptides in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Dulac_neuropeptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuropeptides in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Dulac_neuropeptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Lab Neuropeptides in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Dulac Transcription Factors
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Dulac_TFs <- metadata(PAG_sceset_qc_norm)$genes.DulacTranscriptionfactors[(metadata(PAG_sceset_qc_norm)$genes.DulacTranscriptionfactors %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Dulac_TFs)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Dulac_TFs,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Dulac_TFs,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Dulac_TFs,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Transcription Factors in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Dulac_TFs, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Transcription Factors in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Dulac_TFs, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Transcription Factors in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Dulac_TFs,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Transcription Factors in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Dulac_TFs, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dulac Transcription Factors in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Dulac_TFs,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dulac Transcription Factors in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Yang Dan's Genes
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Dan <- metadata(PAG_sceset_qc_norm)$genes.DanSleep[(metadata(PAG_sceset_qc_norm)$genes.DanSleep %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Dan)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Dan,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Dan,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Dan,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dan Lab Genes in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Dan, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dan Lab Genes in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Dan, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dan Lab Genes in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Dan,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dan Lab Genes in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Dan, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dan Lab Genes in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Dan,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dan Lab Genes in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Cornelius Gross's Genes
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Gross <- metadata(PAG_sceset_qc_norm)$genes.Gross[(metadata(PAG_sceset_qc_norm)$genes.Gross %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Gross)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Gross,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Gross,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Gross,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Gross Lab Genes in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Gross, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Gross Lab Genes in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Gross, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Gross Lab Genes in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Gross,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Gross Lab Genes in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Gross, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Gross Lab Genes in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Gross,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Gross Lab Genes in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Darios's SC Genes
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
PAG_Dario <- metadata(PAG_sceset_qc_norm)$genes.DarioSC[(metadata(PAG_sceset_qc_norm)$genes.DarioSC %in% rownames(PAG_sceset_qc_norm))]
length(PAG_Dario)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, PAG_Dario,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, PAG_Dario,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], PAG_Dario,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dario Genes in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], PAG_Dario, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dario Genes in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], PAG_Dario, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dario Genes in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], PAG_Dario,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dario Genes in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], PAG_Dario, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Dario Genes in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], PAG_Dario,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Dario Genes in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

## Step X.5 | Data Club plots
### Data Club Ion Channels
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
DataClub_ionchannels <- metadata(PAG_sceset_qc_norm)$genes.icDataClub[(metadata(PAG_sceset_qc_norm)$genes.icDataClub %in% rownames(PAG_sceset_qc_norm))]
length(DataClub_ionchannels)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, DataClub_ionchannels,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, DataClub_ionchannels,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], DataClub_ionchannels,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Ion Channels in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], DataClub_ionchannels, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Ion Channels in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], DataClub_ionchannels, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Ion Channels in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], DataClub_ionchannels,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Ion Channels in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], DataClub_ionchannels, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Ion Channels in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], DataClub_ionchannels,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Ion Channels in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

```

### Data Club Neuromodulators
```{r}
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
DataClub_neuromodulators <- metadata(PAG_sceset_qc_norm)$genes.nmDataClub[(metadata(PAG_sceset_qc_norm)$genes.nmDataClub %in% rownames(PAG_sceset_qc_norm))]
length(DataClub_neuromodulators)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, DataClub_neuromodulators,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, DataClub_neuromodulators,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], DataClub_neuromodulators,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuromodulators in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], DataClub_neuromodulators, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuromodulators in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], DataClub_neuromodulators, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuromodulators in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], DataClub_neuromodulators,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuromodulators in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], DataClub_neuromodulators, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuromodulators in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], DataClub_neuromodulators,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuromodulators in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )
```

### Data Club Neuropeptides
```{r} 
# Get the ion channel gene names from the appropriate column and filter any genes not present in our dataset:
DataClub_neuropeptides <- metadata(PAG_sceset_qc_norm)$genes.npDataClub[(metadata(PAG_sceset_qc_norm)$genes.npDataClub %in% rownames(PAG_sceset_qc_norm))]
length(DataClub_neuropeptides)

# Make violing plots
plotExpression(PAG_sceset_qc_norm, DataClub_neuropeptides,  
               x = "cell.type", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 10, show_median = TRUE,
               xlab ="Cell Type")

plotExpression(PAG_sceset_qc_norm, DataClub_neuropeptides,  
               x = "PAG.areacollection", colour_by = "cell.type",
               exprs_values = "logcounts", ncol = 5, show_median = TRUE,
               xlab = "PAG subdivision")

# Make Heatmaps by cell type
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGAT"], DataClub_neuropeptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#ff777c"))(100), border_color = NA, 
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuropeptides in VGAT cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$cell.type=="VGluT2"], DataClub_neuropeptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#009CB5"))(100), border_color = NA,
            colour_columns_by = "PAG.areacollection", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuropeptides in VGluT2 cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

# Make Heatmaps by PAG subdivision
plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dmpag"], DataClub_neuropeptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#F6C324"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuropeptides in dmPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="dlpag"], DataClub_neuropeptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#E3AE09"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuropeptides in dlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="lpag"], DataClub_neuropeptides, 
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#B78C08"))(100), border_color = NA,
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3,
            show_colnames = FALSE,
            main = "Expression of Neuropeptides in lPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

plotHeatmap(PAG_sceset_qc_norm[, PAG_sceset_qc_norm$PAG.areacollection=="vlpag"], DataClub_neuropeptides,
            exprs_values = "logcounts", 
            center = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, #zlim = c(0,10), 
            color = colorRampPalette(c("white", "#8B6A06"))(100), border_color = NA, 
            colour_columns_by = "cell.type", 
            fontsize = 12,# cellwidth = 2, cellheight = 3, 
            show_colnames = FALSE,
            main = "Expression of Neuropeptides in vlPAG cells"#, width = 50, height = 20, filename = "PAG.pdf"
            )

```