---
title: "Topographic, single-cell gene expression profiling of Periaqueductal Gray neurons"
subtitle: "Possibly useful plotting tools"
author:
  - name: "Oriol Pavon Arocas, Sarah F. Olesen, and Tiago Branco"
    affiliation: "Sainsbury Wellcome Centre for Neural Circuits and Behaviour, University College London, UK"
    email: "oriol.pavon.16@ucl.ac.uk"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook:
    highlight: pygments
    number_sections: FALSE
    theme: lumen
    toc: TRUE
    toc_float: TRUE

#output: rmdformats::readthedown:
  #highlight: pygments
---

***

This is a pipeline to analyse single-cell RNA sequencing data from Periaqueductal Gray neurons (1) isolated from acute midbrain slices of transgenic mice using visually guided aspiration via patch pipettes and (2) processed using SMART-seq2 (Picelli et al., Nature Protocols 2014).

This pipeline has been generated after attending the [EMBL-EBI RNA-Sequence Analysis Course](https://www.ebi.ac.uk/training/events/2019/rna-sequence-analysis) and [attending](https://training.csx.cam.ac.uk/bioinformatics/event/2823386) and following the online course on [Analysis of single cell RNA-seq data](https://github.com/hemberg-lab/scRNA.seq.course) by the [Hemberg Lab](https://www.sanger.ac.uk/science/groups/hemberg-group). Many other resources have been used, including the [Orchestrating Single-Cell Analysis with Bioconductor book](https://bioconductor.org/books/release/OSCA/) by Robert Amezquita, Aaron Lun, Stephanie Hicks, and Raphael Gottardo, the [simpleSingleCell workflow in Bioconductor](https://bioconductor.org/packages/3.9/workflows/html/simpleSingleCell.html) maintained by Aaron Lun, the [rnaseqGene workflow](https://bioconductor.org/packages/3.10/workflows/html/rnaseqGene.html) maintained by Michael Love, the [RNAseq123 workflow](https://bioconductor.org/packages/3.10/workflows/html/RNAseq123.html) maintained by Matthew Ritchie, and the [EGSEA123 workflow](https://bioconductor.org/packages/3.10/workflows/html/EGSEA123.html) maintained by Matthew Ritchie.

Other key resources include [Bioconductor](http://www.bioconductor.org/) (Huber et al., Nature Methods 2015), [scRNA-tools](https://www.scrna-tools.org/), `scater` (McCarthy et al., Bioinformatics 2017), `scran` (Lun et al., F1000Res 2016), `SC3` (Kiselev et al., Nature Methods 2017), `Seurat` (Butler et al., Nature Biotechnology 2018), `clusterExperiment` (Risso et al., PLOS Computational Biology 2018), `limma` (Ritchie et al., Nucleic Acids Research 2015), `DESeq2` (Love et al., Genome Biology 2014), `edgeR` (Robinson et al., Bioinformatics 2010), `MAST` (Finak, McDavid, Yajima et al., Genome Biology 2015), `iSEE` (Rue-Albrecht & Marini et al., F1000Research 2018), `t-SNE` (van der Maaten & Hinton, Journal of Machine Learning Research 2008), `UMAP` (McInnes et al., arXiv 2018), and the [Mathematical Statistics and Machine Learning for Life Sciences](https://towardsdatascience.com/tagged/stats-ml-life-sciences) column by Nikolay Oskolkov.

***

# Possibly useful plotting tools
Load the `PAG_sceset_qc_norm_filt_corr` after normalization, filtering, and batch correction. We should thus have a `corrected` slot in `assays`:
```{r}
set.seed(1991)
library(SC3)
library(scater)
library(SingleCellExperiment)
library(pheatmap)
library(mclust)
library(extrafont)

PAG_sceset_qc_norm_filt_corr <- readRDS("PAG_sceset_qc_norm_filt_corr.rds")
assayNames(PAG_sceset_qc_norm_filt_corr)
PAG_sceset_qc_norm_filt_corr
```

## Colour palettes
```{r}
# Colours PAG
colours_celltype_PAGarea <- c(`dmpag_VGAT` = "red", `dlpag_VGAT` = "coral2", `lpag_VGAT` = "salmon", `vlpag_VGAT` = "darksalmon",
                              `dmpag_VGluT2` = "cornflowerblue", `dlpag_VGluT2` = "cyan3", `lpag_VGluT2` = "aquamarine3", `vlpag_VGluT2` = "cadetblue3")

colours_celltype = c(rgb(255, 119, 124, 255/2, maxColorValue = 255), rgb(0, 156, 181, 255/6, maxColorValue = 255))

# Heatmap palettes
color = rev(viridis::magma(100))
colorRampPalette(c("white", "blue"))(100)

# Clustering colours:
cluster_colours_10 <- c(`1` = "#729ECE", `2` = "#FF9E4A", `3` = "#67BF5C", `4` = "#ED665D", `5` = "#AD8BC9",
                        `6` = "#A8786E", `7` = "#ED97CA", `8` = "#A2A2A2", `9` = "#CDCC5D", `10` = "#6DCCDA")

cluster_colours_20 <- c(`1` = "#1F77B4", `2` = "#AEC7E8", `3` = "#FF7F0E", `4` = "#FFBB78", `5` = "#2CA02C",
                        `6` = "#98DF8A", `7` = "#D62728", `8` = "#FF9896", `9` = "#9467BD", `10` = "#C5B0D5",
                        `11` = "#8C564B", `12` = "#C49C94", `13` = "#E377C2", `14` = "#F7B6D2", `15` = "#7F7F7F", 
                        `16` = "#C7C7C7", `17` = "#BCBD22", `18` = "#DBDB8D", `19` = "#17BECF", `20` = "#9EDAE5")

cluster_colors <- scater:::.get_palette("tableau10medium") # hidden scater colours


# Scater palettes:
tableau20 = c("#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78", "#2CA02C",
              "#98DF8A", "#D62728", "#FF9896", "#9467BD", "#C5B0D5",
              "#8C564B", "#C49C94", "#E377C2", "#F7B6D2", "#7F7F7F",
              "#C7C7C7", "#BCBD22", "#DBDB8D", "#17BECF", "#9EDAE5")

tableau10medium = c("#729ECE", "#FF9E4A", "#67BF5C", "#ED665D",
                    "#AD8BC9", "#A8786E", "#ED97CA", "#A2A2A2",
                    "#CDCC5D", "#6DCCDA")

colorblind10 = c("#006BA4", "#FF800E", "#ABABAB", "#595959",
                 "#5F9ED1", "#C85200", "#898989", "#A2C8EC",
                 "#FFBC79", "#CFCFCF")

trafficlight = c("#B10318", "#DBA13A", "#309343", "#D82526",
                 "#FFC156", "#69B764", "#F26C64", "#FFDD71",
                 "#9FCD99")

purplegray12 = c("#7B66D2", "#A699E8", "#DC5FBD", "#FFC0DA",
                 "#5F5A41", "#B4B19B", "#995688", "#D898BA",
                 "#AB6AD5", "#D098EE", "#8B7C6E", "#DBD4C5")

bluered12 = c("#2C69B0", "#B5C8E2", "#F02720", "#FFB6B0", "#AC613C",
              "#E9C39B", "#6BA3D6", "#B5DFFD", "#AC8763", "#DDC9B4",
              "#BD0A36", "#F4737A")

greenorange12 = c("#32A251", "#ACD98D", "#FF7F0F", "#FFB977",
                  "#3CB7CC", "#98D9E4", "#B85A0D", "#FFD94A",
                  "#39737C", "#86B4A9", "#82853B", "#CCC94D")

cyclic = c("#1F83B4", "#1696AC", "#18A188", "#29A03C", "#54A338",
           "#82A93F", "#ADB828", "#D8BD35", "#FFBD4C", "#FFB022",
           "#FF9C0E", "#FF810E", "#E75727", "#D23E4E", "#C94D8C",
           "#C04AA7", "#B446B3", "#9658B1", "#8061B4", "#6F63BB")
```

More colour paletes can be found in the `gameofthrones` package (https://github.com/aljrico/gameofthrones):
```{r}
#install.packages("gameofthrones")
library(gameofthrones)

got_palettes <- list(
  targaryen = c("#000000", "#4e0002", "#9D0005", "#9593A2", "#D8D8D8"),
  targaryen2 = c("#E1DFC4", "#F52156", "#CA001B", "#B30000", "#600000"),
  stark = c("#17181D", "#25252A", "#3C3E43", "#575964", "#9593A2"),
  stark2 = c("#090A0C", "#591D18", "#938170", "#CCA15B", "#636166"),
  stark3 =c("#0B1113",  "#1F2D2F", "#1C262A",  "#2D3A3E", "#989A9C"),
  lannister = c("#5C0000", "#890000", "#C50000", "#FB7E00", "#FFA700"),
  martell = c("#F1B043", "#FF9517", "#EF7300", "#ED4700", "#B20000"),
  tully = c("#131142", "#DADEE1", "#C1000E"),
  greyjoy = c("#000000", "#372E29", "#726255", "#F0C75E", "#ECB939"),
  baratheon = c("#02060B", "#987411", "#B89017", "#F0BA00", "#FEFEF9"),
  baratheon2 = c("#291D15", "#784632", "#8C5A3C", "#966446", "#C88C6E", "#E6D264", "#FFEB78"),
  tyrell = c("#36A107", "#4D852B", "#60AA42", "#C4B92A", "#C4B600"),
  white_walkers = c("#263464", "#345392", "#5492D0", "#6596C9", "#51B2FF"),
  jon_snow = c("#132525", "#113232", "#747374", "#95C2C1", "#36AFAE"),
  margaery = c("#82B5C4", "#A9BC7E", "#D7C42A", "#CDA25C", "#C37E90"),
  daenerys = c("#792427", "#545058", "#2B818E", "#80A098", "#D1BDA2"),
  game_of_thrones = c("#9A1408", "#B32B1B", "#966628", "#AAA549", "#B4AF6D"),
  wildfire = c("#000000", "#002807", "#006300", "#009C00", "#00C800", "#6EFF6E", "#E6FFFF"),
  arya = c("#232D37", "#37414B", "#889999", "#AAB7AF", "#5D7850")
)
```

## SENC Poster dot plot
```{r}
set.seed(1991)
library(tidyverse)
library(scater)
library(SingleCellExperiment)

# Set the directory where your data and scripts are:
setwd("D:/Dropbox (UCL - SWC)/Project_transcriptomics/analysis/PAG_scRNAseq_analysis")

# Set the path to save figures from this Part:
path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_dotplots/"
date <- Sys.Date()
date <- gsub("-", "_", date)

PAG_sceset_qc_norm_filt <- readRDS("PAG_sceset_qc_norm_filt.rds")
PAG_sceset_qc_norm_filt_corr <- readRDS("PAG_sceset_qc_norm_filt_corr.rds")
assayNames(PAG_sceset_qc_norm_filt_corr)
PAG_sceset_qc_norm_filt_corr
```

```{r}
DA_dotplot_vgat <- plotDots(PAG_sceset_qc_norm_filt,
                            features = c("Drd1", "Drd2", "Drd3", "Drd4", "Drd5"),
                            group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                            low_color = "white", high_color = "#FF8080",
                            max_ave = NULL, max_detected = NULL
                            ) + ggtitle("Dopamine")

DA_dotplot_vglut2 <- plotDots(PAG_sceset_qc_norm_filt,
                              features = c("Drd1", "Drd2", "Drd3", "Drd4", "Drd5"),
                              group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                              low_color = "white", high_color = "#0F99B2",
                              max_ave = NULL, max_detected = NULL
                              ) + ggtitle("Dopamine")

DA_dotplot_vgat_subdivisions <- plotDots(PAG_sceset_qc_norm_filt,
                                         features = c("Drd1", "Drd2", "Drd3", "Drd4", "Drd5"),
                                         group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                         low_color = "white", high_color = "#FF8080",
                                         max_ave = NULL, max_detected = NULL
                                         ) + ggtitle("Dopamine")

DA_dotplot_vglut2_subdivisions <- plotDots(PAG_sceset_qc_norm_filt,
                                           features = c("Drd1", "Drd2", "Drd3", "Drd4", "Drd5"),
                                           group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                           low_color = "white", high_color = "#0F99B2",
                                           max_ave = NULL, max_detected = NULL
                                           ) + ggtitle("Dopamine")

ggsave(filename = str_c(date, "_DA_dotplot_vgat.pdf"),
       plot = DA_dotplot_vgat,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 9, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_DA_dotplot_vglut2.pdf"),
       plot = DA_dotplot_vglut2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 9, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_DA_dotplot_vgat_subdivisions.pdf"),
       plot = DA_dotplot_vgat_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 9, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_DA_dotplot_vglut2_subdivisions.pdf"),
       plot = DA_dotplot_vglut2_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 9, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

```{r}
serotonin_receptors <- c("Htr1a", "Htr1b", "Htr1d", "Htr1e", "Htr1f", "Htr2a", "Htr2b", "Htr2c", "Htr3a", "Htr3b", "Htr3c", "Htr3d", "Htr3e", "Htr4", "Htr5a", "Htr6", "Htr7")
serotonin_receptors <- serotonin_receptors[(serotonin_receptors %in% rownames(PAG_sceset_qc_norm_filt))]

plotDots(PAG_sceset_qc_norm_filt,
         features = serotonin_receptors,
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Serotonin")

plotDots(PAG_sceset_qc_norm_filt,
         features = serotonin_receptors,
         group = "celltype_PAGarea",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Serotonin")

plotDots(PAG_sceset_qc_norm_filt,
         features = c("Adra1a", "Adra1b", "Adra1d", "Adra2a", "Adra2b", "Adra2c", "Adrb1", "Adrb2", "Adrb3"),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Adrenergic receptors")

plotDots(PAG_sceset_qc_norm_filt,
         features = c("Adra1a", "Adra1b", "Adra1d", "Adra2a", "Adra2b", "Adra2c", "Adrb1", "Adrb2", "Adrb3"),
         group = "celltype_PAGarea",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Adrenergic receptors")
```


## scater::plotDot()
Create a dot plot of expression values for a grouping of cells, where the size and color of each dot represents the proportion of detected expression values and the average expression, respectively, for each feature in each group of cells.
```{r}
library(scater)
#names(metadata(PAG_sceset_qc_norm_filt_corr_clust_de))

plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features = c("Drd1", "Drd2", "Drd3", "Drd4", "Drd5", "Dbh", "Th", "Ddc"),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Dopamine")

plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features = c("Htr1a", "Htr1b", "Htr1d", "Htr1e", "Htr1f",
                      "Htr2a", "Htr2b", "Htr2c",
                      "Htr3a", "Htr3b", "Htr3c", "Htr3d", "Htr3e",
                      "Htr4", "Htr5a", "Htr6", "Htr7"),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Serotonin")


plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features = c("Adra1a", "Adra1b", "Adra1d", "Adra2a", "Adra2b", "Adra2c", "Adrb1", "Adrb2", "Adrb3"),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Adrenergic receptors")

plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features = c("Asic4", "Pnoc","Cacna2d1", "Tac1"),
         group = "celltype_PAGarea",
         low_color = "white", high_color = "red",
         max_ave = 5, max_detected = NULL
         ) + ggtitle("Some Differentially expressed genes")

plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features = c("Kcnj2", "Kcnj12", "Kcnj4", "Kcnj14", "Kcnj3", "Kcnj6", "Kcnj9", "Kcnj5", "Kcnj10", "Kcnj16", "Kcnj8", "Kcnj11", "Kcnj13"),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("IRK channels")

plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features = c("Kcnj3", "Kcnj6", "Kcnj9", "Kcnj5"),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("GIRK channels")

pacemaker_channels <- c("Hcn1", "Hcn2", "Hcn3", "Hcn4", "Kcna4", "Kcnc3", "Kcnc4", "Kcnd1", "Kcnd2", "Kcnd3", "Kcnj6", "Scn7a", "Scn8a", "Scn9a", "Cacna1g", "Cacna1h", "Cacna1i", "Cacna2d1")
pacemaker_channels <- pacemaker_channels[(pacemaker_channels %in% rownames(PAG_sceset_qc_norm_filt_corr_clust_de))]
plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features =  pacemaker_channels,
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Pacemaking ion channels")

# Ion Channels
plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features =  metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.npDataClub[(metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.npDataClub %in% rownames(PAG_sceset_qc_norm_filt_corr_clust_de))],
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Ion Channels")

plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features =  metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.npDataClub[(metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.npDataClub %in% rownames(PAG_sceset_qc_norm_filt_corr_clust_de))],
         group = "celltype_PAGarea",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Ion Channels")

# DE genes
plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features =  c(VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.ionchannels],
                       VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.ionchannels]),
         group = "cell.type",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("DE Ion channels")

# Monoamines
plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
         features =  metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Monoamines[(metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Monoamines %in% rownames(PAG_sceset_qc_norm_filt_corr_clust_de))],
         group = "celltype_PAGarea",
         low_color = "white", high_color = "red",
         max_ave = NULL, max_detected = NULL
         ) + ggtitle("Neuromodulators")
```

## Dot plots - Full control
https://ltla.github.io/SingleCellThoughts/general/visualization.html
```{r}
library(scRNAseq)
sce <- ZeiselBrainData()
sce <- logNormCounts(sce)
library(scater)
ids <- colData(sce)[,c("level1class", "tissue")] # cell type and group.
my.genes <- sample(rownames(sce), 5) # pick 5 genes, or your markers, whatever.
# Compute average log-count per tissue/cell type combination.
# Also compute proportion of cells with detected expression.
ave <- sumCountsAcrossCells(sce, ids=ids, exprs_values='logcounts', average=TRUE, subset_row=my.genes)
prop <- numDetectedAcrossCells(sce, ids=ids, average=TRUE, subset_row=my.genes)
# Melt this into a long-form data.frame.
combined <- melt(assay(ave))
colnames(combined) <- c("Gene", "Column", "Average")
combined$Proportion <- melt(assay(prop))[,3]
combined <- cbind(combined, colData(ave)[combined$Column,])
# Profit.
ggplot(combined, mapping=aes(x=Gene, y=level1class)) +
    geom_point(mapping=aes(size=Proportion, colour=Average)) +
    scale_color_gradient(low="white", high="red") +
    facet_wrap(~tissue)
```


## Patchwork - composing ggplots
https://patchwork.data-imaginist.com/
```{r}
library(ggplot2)
library(patchwork)

p1 <- ggplot(mtcars) + geom_point(aes(mpg, disp))
p2 <- ggplot(mtcars) + geom_boxplot(aes(gear, disp, group = gear))

p1 + p2

p3 <- ggplot(mtcars) + geom_smooth(aes(disp, qsec))
p4 <- ggplot(mtcars) + geom_bar(aes(carb))

(p1 | p2 | p3) /
      p4
```

## gganimate - animating ggplots
https://gganimate.com/
```{r}
library(ggplot2)
library(gganimate)

ggplot(mtcars, aes(factor(cyl), mpg)) + 
  geom_boxplot() + 
  # Here comes the gganimate code
  transition_states(
    gear,
    transition_length = 2,
    state_length = 1
  ) +
  enter_fade() + 
  exit_shrink() +
  ease_aes('sine-in-out')
```

## Allen atlas plots
https://github.com/AllenInstitute/scrattch.vis/
```{r}
library(tasic2016data)
library(scrattch.vis)
options(stringsAsFactors = F)

anno <- tasic_2016_anno
anno <- anno[anno$primary_type_id > 0,]
data <- tasic_2016_rpkm
data_df <- cbind(sample_name = colnames(data),
                 as.data.frame(t(data[c("Pvalb","Sst","Rorb"),])))

group_violin_plot(data_df, 
                  anno, 
                  genes = c("Pvalb","Sst","Rorb"), 
                  grouping = "primary_type", 
                  log_scale = FALSE,
                  font_size = 5,
                  rotate_counts = TRUE)
```

## Make a final figure
https://github.com/massonix/sampling_artifacts/blob/master/1-PBMC/08-gene_signature_Smartseq.Rmd
```{r}
# Plot tSNE with k = 2
tsne_k2 <- plot_tsne(sce = sce,
                     exprs_value = "logcounts",
                     color_by = "sc3_2_clusters",
                     point_size = 2.5,
                     point_alpha = 1,
                     colors = c("royalblue1", "red2"),
                     title = "SC3 clusters (k=2)"
                     )
tsne_k2 <- tsne_k2 +
  scale_color_manual(values = c("royalblue1", "red2"), labels = c("cluster 1", "cluster 2")) +
  theme(legend.title = element_blank())
# Save tSNE
ggsave(plot = tsne_k2, 
       filename = str_c("results/plots/", date, "_Smart-seq2_tsne_k2.pdf"), 
       device = "pdf",
       height = 7, 
       width = 8
       )
tsne_k2
saveRDS(object = sce, file = "results/R_objects/sce_Smart-seq2.rds")

# Effect of time + explained variability
set.seed(1)
palette <- c("#999999", "#92e8df", "#632c63", "#e4624e", "#c0e212")
sce_f <- scater::filter(sce, temperature != "4ºC")
tsne_smart_time <- plot_tsne(sce = sce_f,
                             exprs_value = "logcounts",
                             color_by = "time",
                             point_size = 3,
                             point_alpha = 1,
                             colors = palette,
                             title = "Smart-seq2"
                             )

tsne_smart_time <- tsne_smart_time + theme(legend.position = "bottom")
legend_tsne_smart <- as_ggplot(get_legend(tsne_smart_time))
tsne_smart_time <- tsne_smart_time + theme(plot.title = element_blank(),
                                           legend.position = "none",
                                           plot.background = element_blank(),
                                           panel.border = element_blank(),
                                           axis.title = element_blank(),
                                           axis.ticks = element_blank(),
                                           axis.text = element_blank())
legend_explained_var <- as_ggplot(get_legend(explained_var))
explained_var <- explained_var + theme(legend.position = "none")

# SC3 clusters + distribution 0h/RT/4ºC cells in each cluster
legend_k2 <- as_ggplot(get_legend(cluster_distr_gg))
cluster_distr_gg <- cluster_distr_gg + theme(legend.position = "none")
tsne_k2 <- tsne_k2 + theme(legend.position = "bottom")
legend_tsne_k2 <- as_ggplot(get_legend(tsne_k2))
tsne_k2 <- tsne_k2 + theme(plot.title = element_blank(),
                           legend.position = "none",
                           plot.background = element_blank(),
                           panel.border = element_blank(),
                           axis.title = element_blank(),
                           axis.ticks = element_blank(),
                           axis.text = element_blank())

# Violin plot + tsne regressed
violin_smart <- violin_smart + theme(legend.position = "none") + ggpubr::rotate_x_text(angle = -45, hjust = 0)
tsne_regressed <- tsne_regressed + theme(legend.position = "bottom") 
legend_tsne_regressed <- as_ggplot(get_legend(tsne_regressed))
tsne_regressed <- tsne_regressed + theme(plot.title = element_text(face = "bold"),
                                         legend.position = "none",
                                         plot.background = element_blank(),
                                         panel.border = element_blank(),
                                         axis.title = element_blank(),
                                         axis.ticks = element_blank(),
                                         axis.text = element_blank())

# Arrange figure
tsne_smart_explained <- plot_grid( tsne_smart_time, 
                                   NULL,
                                   explained_var,
                                   ncol = 3, 
                                   nrow = 1,
                                   rel_widths = c(1, 0.1, 1),
                                   labels = c("a", "b"),
                                   align = "h",
                                   axis = "b"
                                   )
k2_distr <- plot_grid(tsne_k2, 
                      NULL,
                      cluster_distr_gg,
                      ncol = 3, 
                      nrow = 1,
                      rel_widths = c(1, 0.1, 1),
                      labels = c("c", "d"),
                      align = "h",
                      axis = "b"
                      )
violin_regressed <- plot_grid(violin_smart, 
                              NULL,
                              tsne_regressed,
                              ncol = 3, 
                              nrow = 1,
                              rel_widths = c(1, 0.1, 1),
                              labels = c("e", "f"),
                              align = "h",
                              axis = "b"
                              )
smart_fig <- plot_grid(tsne_smart_explained, 
                       NULL,
                       k2_distr,
                       violin_regressed,
                       ncol = 1, 
                       nrow = 4,
                       rel_heights = c(1, 0.05, 1, 1)
                       )
# Save figure
ggsave(filename = str_c("doc/figures/R/", date, "_smart_seq_figure.pdf"), 
       plot = smart_fig, 
       width = 18, 
       height = 26,
       units = "cm"
       )
# Save legends
legends_list <- list(legend_tsne_smart, legend_explained_var, legend_k2, legend_tsne_k2, legend_tsne_regressed)
names(legends_list) <- c("legend_tsne_smart", "legend_explained_var", "legend_k2", "legend_tsne_k2", "legend_tsne_regressed")
walk(names(legends_list), function(leg) {
  ggsave(
    filename = str_c("doc/figures/legends/", date, "_", leg, ".pdf"), 
    plot = legends_list[[leg]], 
    width = 9, 
    height = 5,
    units = "cm"
  )
})

# Differential Expression Analysis
Idents(pbmc_female) <- "temperature"
pbmc_f_sub <- subset(pbmc_female, idents = c("fresh", "4ºC"))
dea <- FindMarkers(pbmc_f_sub, ident.1 = "4ºC", test.use = "wilcox")
dea <- rownames_to_column(dea, var = "gene")
dea <- dplyr::mutate(dea, significance = ifelse(p_val_adj < 0.001, "sig", "no sig"))
subset_data <- dplyr::filter(dea, avg_logFC > 0.9)
volcano <- dea %>% 
  dplyr::mutate(significance = ifelse(p_val_adj < 0.001, "sig", "no sig")) %>% 
  ggplot(aes(avg_logFC, -1 * log10(p_val_adj), color = significance)) +
    geom_point() +
    geom_text_repel(data = subset_data, aes(label = gene), color = "black", size = 2.5) +
    scale_color_manual(values = c("gray78", "chartreuse2")) +
    labs(x = "log (fold-change)", y = "-log10 (p-value)", color = "") +
    theme_classic()
volcano
saveRDS(volcano, "results/R_objects/ggplots/volcano_4ºC_pbmc.rds")
```

# Building figures
## Plots from Part 2
```{r}
library(tidyverse)
library(ggplot2)
library(patchwork)

path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part2_dataQC/"
date <- Sys.Date()
date <- gsub("-", "_", date)
```

Violin plots from transgenes:
```{r}
library(patchwork)
# Load violin plots for fluorophores
violin_fluorophores_EYFPbymouse_counts <- readRDS(file = str_c(path_for_figures, "violin_fluorophores_EYFPbymouse_counts.rds"))
violin_fluorophores_EYFPbymouse_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_fluorophores_EYFPbymouse_logcounts_raw.rds"))
violin_fluorophores_tdTomatobymouse_counts <- readRDS(file = str_c(path_for_figures, "violin_fluorophores_tdTomatobymouse_counts.rds"))
violin_fluorophores_tdTomatobymouse_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_fluorophores_tdTomatobymouse_logcounts_raw.rds"))
violin_fluorophores_celltype_counts <- readRDS(file = str_c(path_for_figures, "violin_fluorophores_celltype_counts.rds"))
violin_fluorophores_celltype_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_fluorophores_celltype_logcounts_raw.rds"))

# Refreshment of how these plots were created
# theme_args <- theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
# violin_fluorophores_EYFPbymouse_logcounts_raw <- plotExpression(PAG_sceset[, PAG_sceset$cell.fluorophore == "EYFP"], c("EYFP", "tdTomato", "Cre"), x = "mouse.id",
#                                                                 exprs_values = "logcounts_raw",
#                                                                 colour_by = "batch.processing", ncol = 3, theme_size = 11, point_alpha = 0.8) +
#   ggtitle("Transgene expression levels in cells from Cre::EYFP animals (log2 raw counts)") + xlab("Mouse ID") + ylab("Expression (log2 raw counts)") + theme_args
# violin_fluorophores_celltype_counts <- plotExpression(PAG_sceset, c("EYFP", "tdTomato"), x = "cell.fluorophore",
#                                                       exprs_values = "counts", 
#                                                       colour_by = "mouse.id", ncol = 2, theme_size = 11, point_alpha = 0.8) +
#   ggtitle("Transporter expression levels in cells from VGluT2::Cre animals (raw counts)") + xlab("Fluorophore") + ylab("Expression (raw counts)")

# Fix individual plots
violin_fluorophores_EYFPbymouse_counts <- violin_fluorophores_EYFPbymouse_counts + ggtitle("Cre::EYFP animals")+ xlab("Mouse ID") + ylab("raw counts") 
violin_fluorophores_EYFPbymouse_logcounts_raw <- violin_fluorophores_EYFPbymouse_logcounts_raw + ggtitle("Cre::EYFP animals")+ xlab("Mouse ID") + ylab("log2 raw counts") 
violin_fluorophores_tdTomatobymouse_counts <- violin_fluorophores_tdTomatobymouse_counts + ggtitle("Cre::tdTomato animals")+ xlab("Mouse ID") + ylab("raw counts") 
violin_fluorophores_tdTomatobymouse_logcounts_raw <- violin_fluorophores_tdTomatobymouse_logcounts_raw + ggtitle("Cre::tdTomato animals")+ xlab("Mouse ID") + ylab("log2 raw counts") 
violin_fluorophores_celltype_counts <- violin_fluorophores_celltype_counts + ggtitle(NULL)+ xlab("Fluorophore") + ylab("raw counts") 
violin_fluorophores_celltype_logcounts_raw <- violin_fluorophores_celltype_logcounts_raw + ggtitle(NULL)+ xlab("Fluorophore") + ylab("log2 raw counts") 

# Compose figures using `patchwork`
violin_fluorophores_byanimal_composed <- (violin_fluorophores_EYFPbymouse_counts + violin_fluorophores_EYFPbymouse_logcounts_raw) / 
  (violin_fluorophores_tdTomatobymouse_counts + violin_fluorophores_tdTomatobymouse_logcounts_raw) + 
  plot_annotation(title = "Transgene expression levels", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

violin_fluorophores_bycelltype_composed <- (violin_fluorophores_celltype_counts / violin_fluorophores_celltype_logcounts_raw) + 
  plot_annotation(title = "Fluorophore expression levels", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

# Plot composed figures
violin_fluorophores_byanimal_composed
violin_fluorophores_bycelltype_composed

# Save composed figures
ggsave(filename = str_c(date, "_composed_violin_fluorophores_byanimal.pdf"),
       plot = violin_fluorophores_byanimal_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 24, height = 12, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )

ggsave(filename = str_c(date, "_composed_violin_fluorophores_bycelltype.pdf"),
       plot = violin_fluorophores_bycelltype_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 12, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

Violin plots from transporters:
```{r}
library(patchwork)
# Load violin plots for transporters
violin_VGATVGluT2_VGATbymouse_counts <- readRDS(file = str_c(path_for_figures, "violin_VGATVGluT2_VGATbymouse_counts.rds"))
violin_VGATVGluT2_VGATbymouse_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_VGATVGluT2_VGATbymouse_logcounts_raw.rds"))
violin_VGATVGluT2_VGluT2bymouse_counts <- readRDS(file = str_c(path_for_figures, "violin_VGATVGluT2_VGluT2bymouse_counts.rds"))
violin_VGATVGluT2_VGluT2bymouse_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_VGATVGluT2_VGluT2bymouse_logcounts_raw.rds"))
violin_VGATVGluT2_celltype_counts <- readRDS(file = str_c(path_for_figures, "violin_VGATVGluT2_celltype_counts.rds"))
violin_VGATVGluT2_celltype_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_VGATVGluT2_celltype_logcounts_raw.rds"))

# Refreshment of how these plots were created
# theme_args <- theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
# violin_fluorophores_EYFPbymouse_logcounts_raw <- plotExpression(PAG_sceset[, PAG_sceset$cell.fluorophore == "EYFP"], c("EYFP", "tdTomato", "Cre"), x = "mouse.id",
#                                                                 exprs_values = "logcounts_raw",
#                                                                 colour_by = "batch.processing", ncol = 3, theme_size = 11, point_alpha = 0.8) +
#   ggtitle("Transgene expression levels in cells from Cre::EYFP animals (log2 raw counts)") + xlab("Mouse ID") + ylab("Expression (log2 raw counts)") + theme_args
# violin_fluorophores_celltype_counts <- plotExpression(PAG_sceset, c("EYFP", "tdTomato"), x = "cell.fluorophore",
#                                                       exprs_values = "counts", 
#                                                       colour_by = "mouse.id", ncol = 2, theme_size = 11, point_alpha = 0.8) +
#   ggtitle("Transporter expression levels in cells from VGluT2::Cre animals (raw counts)") + xlab("Fluorophore") + ylab("Expression (raw counts)")

# Fix individual plots
violin_VGATVGluT2_VGATbymouse_counts <- violin_VGATVGluT2_VGATbymouse_counts + ggtitle("VGAT::Cre animals")+ xlab("Mouse ID") + ylab("raw counts") 
violin_VGATVGluT2_VGATbymouse_logcounts_raw <- violin_VGATVGluT2_VGATbymouse_logcounts_raw + ggtitle("VGAT::Cre  animals")+ xlab("Mouse ID") + ylab("log2 raw counts") 
violin_VGATVGluT2_VGluT2bymouse_counts <- violin_VGATVGluT2_VGluT2bymouse_counts + ggtitle("VGluT2::Cre  animals")+ xlab("Mouse ID") + ylab("raw counts") 
violin_VGATVGluT2_VGluT2bymouse_logcounts_raw <- violin_VGATVGluT2_VGluT2bymouse_logcounts_raw + ggtitle("VGluT2::Cre animals")+ xlab("Mouse ID") + ylab("log2 raw counts") 
violin_VGATVGluT2_celltype_counts <- violin_VGATVGluT2_celltype_counts + ggtitle(NULL)+ xlab("Cell type") + ylab("raw counts") 
violin_VGATVGluT2_celltype_logcounts_raw <- violin_VGATVGluT2_celltype_logcounts_raw + ggtitle(NULL)+ xlab("Cell type") + ylab("log2 raw counts") 

# Compose figures using `patchwork`
violin_transporters_byanimal_composed <- (violin_VGATVGluT2_VGATbymouse_counts + violin_VGATVGluT2_VGATbymouse_logcounts_raw) / 
  (violin_VGATVGluT2_VGluT2bymouse_counts + violin_VGATVGluT2_VGluT2bymouse_logcounts_raw) + 
  plot_annotation(title = "VGAT-VGluT2 expression levels", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

violin_transporters_bycelltype_composed <- (violin_VGATVGluT2_celltype_counts / violin_VGATVGluT2_celltype_logcounts_raw) + 
  plot_annotation(title = "VGAT-VGluT2 expression levels", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))


# Plot composed figures
violin_transporters_byanimal_composed
violin_transporters_bycelltype_composed

# Save composed figures
ggsave(filename = str_c(date, "_composed_violin_transporters_bycelltype.pdf"),
       plot = violin_transporters_byanimal_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 20, height = 10, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )

ggsave(filename = str_c(date, "_composed_violin_transporters_bycelltype.pdf"),
       plot = violin_transporters_bycelltype_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 12, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

Violin plots from QC filters:
```{r}
library(patchwork)
# Load individual violin plots, highlighting excluded cells:
violin_cell_QC_library <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_library.rds"))
violin_cell_QC_detected <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_detected.rds"))
violin_cell_QC_mitochondrial <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_mitochondrial.rds"))
violin_cell_QC_ribosomal <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_ribosomal.rds"))
violin_cell_QC_ERCC <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_ERCC.rds"))
violin_cell_QC_TSO <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_TSO.rds"))
violin_cell_QC_top100genes <-readRDS(file = str_c(path_for_figures, "violin_cell_QC_top100genes.rds"))
violin_cell_QC_EYFP <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_EYFP.rds"))
violin_cell_QC_tdTomato <- readRDS(file = str_c(path_for_figures, "violin_cell_QC_tdTomato.rds"))

# Refreshment of how these plots were created
# violin_cell_QC_ERCC <- plotColData(PAG_sceset, x = "cell.type", y = "altexps_ERCC_percent",
#                                    colour_by = "filter_by_pct_ERCC_manual", theme_size = 11, point_alpha = 0.8) + 
#   ggtitle("Percent reads in ERCC spike-ins") + xlab(NULL) + ylab("% reads in ERCC spike-ins") + 
#   scale_colour_manual(name = "Excluded", values = c("grey", "red"), aesthetics = c("colour", "fill"))

# Fix individual plots
violin_cell_QC_library <- violin_cell_QC_library + ggtitle("Library size") + ylab("Total reads (millions)")
violin_cell_QC_detected <- violin_cell_QC_detected + ggtitle("Detected genes") + ylab("Number of genes detected")
violin_cell_QC_mitochondrial <- violin_cell_QC_mitochondrial + ggtitle("Mitochondrial genes") + ylab("% reads")
violin_cell_QC_ribosomal <- violin_cell_QC_ribosomal + ggtitle("Ribosomal genes") + ylab("% reads")
violin_cell_QC_ERCC <- violin_cell_QC_ERCC + ggtitle("ERCC spike-ins") + ylab("% reads")
violin_cell_QC_TSO <- violin_cell_QC_TSO + ggtitle("TSO concatamers") + ylab("% reads")
violin_cell_QC_top100genes <- violin_cell_QC_top100genes + ggtitle("Top 100 expressed genes") + ylab("% reads")
violin_cell_QC_EYFP <- violin_cell_QC_EYFP + ggtitle("EYFP expression") + ylab("log2 raw counts")
violin_cell_QC_tdTomato <- violin_cell_QC_tdTomato + ggtitle("tdTomato expression") + ylab("log2 raw counts")

# Compose figure using `patchwork`
cell_QC_summary_composed <- (violin_cell_QC_library + violin_cell_QC_detected + violin_cell_QC_mitochondrial) / 
  (violin_cell_QC_ribosomal + violin_cell_QC_ERCC + violin_cell_QC_TSO) /
  (violin_cell_QC_top100genes + violin_cell_QC_EYFP + violin_cell_QC_tdTomato) +
  plot_annotation(title = "Summary of cell QC filters", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

# Plot composed figures
cell_QC_summary_composed

# Save composed figures
ggsave(filename = str_c(date, "_composed_cell_QC_summary.pdf"),
       plot = cell_QC_summary_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 10, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

## Plots from Part 3
```{r}
library(tidyverse)
library(ggplot2)
library(patchwork)

path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part3_normalization/"
date <- Sys.Date()
date <- gsub("-", "_", date)
```

Violin plots from transporters:
```{r}
library(patchwork)

# Load violin plots for transporters
violin_QC_VGATbymouse_VGATVGluT2_counts <- readRDS(file = str_c(path_for_figures, "violin_QC_VGATbymouse_VGATVGluT2_counts.rds"))
violin_QC_VGATbymouse_VGATVGluT2_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_QC_VGATbymouse_VGATVGluT2_logcounts_raw.rds"))
violin_QC_VGATbymouse_VGATVGluT2_logcounts <- readRDS(file = str_c(path_for_figures, "violin_QC_VGATbymouse_VGATVGluT2_logcounts.rds"))
violin_QC_VGluT2bymouse_VGATVGluT2_counts <- readRDS(file = str_c(path_for_figures, "violin_QC_VGluT2bymouse_VGATVGluT2_counts.rds"))
violin_QC_VGluT2bymouse_VGATVGluT2_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_QC_VGluT2bymouse_VGATVGluT2_logcounts_raw.rds"))
violin_QC_VGluT2bymouse_VGATVGluT2_logcounts <- readRDS(file = str_c(path_for_figures, "violin_QC_VGluT2bymouse_VGATVGluT2_logcounts.rds"))
violin_QC_celltype_VGATVGluT2_counts <- readRDS(file = str_c(path_for_figures, "violin_QC_celltype_VGATVGluT2_counts.rds"))
violin_QC_celltype_VGATVGluT2_logcounts_raw <- readRDS(file = str_c(path_for_figures, "violin_QC_celltype_VGATVGluT2_logcounts_raw.rds"))
violin_QC_celltype_VGATVGluT2_logcounts <- readRDS(file = str_c(path_for_figures, "violin_QC_celltype_VGATVGluT2_logcounts.rds"))

# Refreshment of how these plots were created
# theme_args <- theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
# violin_QC_VGATbymouse_VGATVGluT2_counts <- plotExpression(PAG_sceset_qc[, PAG_sceset_qc$cell.type == "VGAT"], c("Slc32a1", "Slc17a6"), x = "mouse.id",
#                                                           exprs_values = "counts",
#                                                           colour_by = "batch.processing", ncol = 2, theme_size = 11, point_alpha = 0.8) +
#   ggtitle("Transporter expression levels in cells from VGAT::Cre animals (raw counts)") + xlab("Mouse ID") + ylab("Expression (raw counts)") + theme_args
# violin_QC_celltype_VGATVGluT2_counts <- plotExpression(PAG_sceset_qc, c("Slc32a1", "Slc17a6"), x = "cell.type",
#                                                        exprs_values = "counts", 
#                                                        colour_by = "mouse.id", ncol = 2, theme_size = 11, point_alpha = 0.8) +
#   ggtitle("Transporter expression levels (raw counts)") + xlab("Cell type") + ylab("Expression (raw counts)")

# Fix individual plots
violin_QC_VGATbymouse_VGATVGluT2_counts <- violin_QC_VGATbymouse_VGATVGluT2_counts + ggtitle("VGAT::Cre animals")+ xlab("Mouse ID") + ylab("raw counts")
violin_QC_VGATbymouse_VGATVGluT2_logcounts_raw <- violin_QC_VGATbymouse_VGATVGluT2_logcounts_raw + ggtitle("VGAT::Cre animals")+ xlab("Mouse ID") + ylab("log2 raw counts")
violin_QC_VGATbymouse_VGATVGluT2_logcounts <- violin_QC_VGATbymouse_VGATVGluT2_logcounts + ggtitle("VGAT::Cre animals")+ xlab("Mouse ID") + ylab("logcounts")
violin_QC_VGluT2bymouse_VGATVGluT2_counts <- violin_QC_VGluT2bymouse_VGATVGluT2_counts + ggtitle("VGluT2::Cre animals")+ xlab("Mouse ID") + ylab("raw counts")
violin_QC_VGluT2bymouse_VGATVGluT2_logcounts_raw <- violin_QC_VGluT2bymouse_VGATVGluT2_logcounts_raw + ggtitle("VGluT2::Cre animals")+ xlab("Mouse ID") + ylab("log2 raw counts")
violin_QC_VGluT2bymouse_VGATVGluT2_logcounts <- violin_QC_VGluT2bymouse_VGATVGluT2_logcounts + ggtitle("VGluT2::Cre animals")+ xlab("Mouse ID") + ylab("logcounts")
violin_QC_celltype_VGATVGluT2_counts <- violin_QC_celltype_VGATVGluT2_counts + ggtitle(NULL)+ xlab("Cell type") + ylab("raw counts")
violin_QC_celltype_VGATVGluT2_logcounts_raw <- violin_QC_celltype_VGATVGluT2_logcounts_raw + ggtitle(NULL)+ xlab("Cell type") + ylab("log2 raw counts")
violin_QC_celltype_VGATVGluT2_logcounts <- violin_QC_celltype_VGATVGluT2_logcounts + ggtitle(NULL)+ xlab("Cell type") + ylab("logcounts")

# Compose figures using `patchwork`
violin_QC_transporters_byanimal_composed <- (violin_QC_VGATbymouse_VGATVGluT2_counts + violin_QC_VGATbymouse_VGATVGluT2_logcounts_raw + violin_QC_VGATbymouse_VGATVGluT2_logcounts) / 
  (violin_QC_VGluT2bymouse_VGATVGluT2_counts + violin_QC_VGluT2bymouse_VGATVGluT2_logcounts_raw + violin_QC_VGluT2bymouse_VGATVGluT2_logcounts) + 
  plot_annotation(title = "VGAT-VGluT2 expression levels after QC", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

violin_QC_transporters_bycelltype_composed <- (violin_QC_celltype_VGATVGluT2_counts / violin_QC_celltype_VGATVGluT2_logcounts_raw + violin_QC_celltype_VGATVGluT2_logcounts) + 
  plot_annotation(title = "VGAT-VGluT2 expression levels after QC", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

# Plot composed figures
violin_QC_transporters_byanimal_composed
violin_QC_transporters_bycelltype_composed

# Save composed figures
ggsave(filename = str_c(date, "_composed_violin_QC_transporters_byanimal.pdf"),
       plot = violin_QC_transporters_byanimal_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 30, height = 10, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )

ggsave(filename = str_c(date, "_composed_violin_QC_transporters_bycelltype.pdf"),
       plot = violin_QC_transporters_bycelltype_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 12, height = 18, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

## Plots from Part 4
```{r}
library(tidyverse)
library(ggplot2)
library(patchwork)

path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part4_HVGs/"
date <- Sys.Date()
date <- gsub("-", "_", date)
```

```{r}
library(patchwork)

# Load violin plots for HVGs
violin_HVG_var <- readRDS(file = str_c(path_for_figures, "violin_HVG_var.rds"))
violin_HVG_var_block <- readRDS(file = str_c(path_for_figures, "violin_HVG_var_block.rds"))
violin_HVG_cv2 <- readRDS(file = str_c(path_for_figures, "violin_HVG_cv2.rds"))
violin_HVG_cv2_block <- readRDS(file = str_c(path_for_figures, "violin_HVG_cv2_block.rds"))

# Refreshment of how these plots were created
# theme_args <- theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1), axis.ticks.x = element_blank())
# violin_HVG_var <- plotExpression(PAG_sceset_qc_norm_filt, features = rownames(hvg_var_out_no_spikes_filt)[1:50],
#                                  exprs_values = "logcounts", theme_size = 11, point_alpha = 0.8
#                                  ) + ggtitle("Top 50 HVGs from mean-var modelling") + theme_args

# Fix individual plots
violin_HVG_var <- violin_HVG_var + ggtitle("Mean-var modelling")
violin_HVG_var_block <- violin_HVG_var_block + ggtitle("Mean-var modelling blocking by batch")
violin_HVG_cv2 <- violin_HVG_cv2 + ggtitle("Mean-cv2 modelling")
violin_HVG_cv2_block <- violin_HVG_cv2_block + ggtitle("Mean-cv2 modelling blocking by batch")

# Compose figures using `patchwork`
violin_HVG_composed <- violin_HVG_var / violin_HVG_var_block / violin_HVG_cv2 / violin_HVG_cv2_block + 
  plot_annotation(title = "Top 50 Highly Variable Genes", theme = theme(plot.title = element_text(size = 18)), tag_levels = "A") + 
  plot_layout(guides = "collect") + theme(plot.tag = element_text(size = 14))

# Plot composed figures
violin_HVG_composed

# Save composed figures
ggsave(filename = str_c(date, "_composed_violin_HVGs.pdf"),
       plot = violin_HVG_composed,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = path_for_figures,
       width = 20, height = 20, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

## Plots from Part 5
```{r}
library(tidyverse)
library(ggplot2)
library(patchwork)

path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part5_batchCorr_reducedDim/"
date <- Sys.Date()
date <- gsub("-", "_", date)
```

Explanatory variables:
```{r}
library(patchwork)

# Load plots from plotExplanatoryVariables()
explanatory_variables_counts <- readRDS(file = str_c(path_for_figures, "explanatory_variables_counts.rds"))
explanatory_variables_logcounts_raw <- readRDS(file = str_c(path_for_figures, "explanatory_variables_logcounts_raw.rds"))
explanatory_variables_logcounts <- readRDS(file = str_c(path_for_figures, "explanatory_variables_logcounts.rds"))
explanatory_variables_corrected <- readRDS(file = str_c(path_for_figures, "explanatory_variables_corrected.rds"))

# Refreshment of how these plots were created
# explanatory_variables_corrected <- plotExplanatoryVariables(PAG_sceset_qc_norm_filt,
#                                                             nvars_to_plot = 17,
#                                                             exprs_values = "corrected",
#                                                             variables = c("mouse.id", "mouse.sex", "mouse.age", "mouse.singlehousedays",
#                                                                           "cell.type", "cell.fluorophore", "slice.number", "slice.depth", "time.sinceslicinghour",
#                                                                           "PAG.area", "PAG.hemisphere", "PAG.PaxinosAPmanual", "celltype_PAGarea",
#                                                                           "batch.processing", "batch.sequencing_round", "detected", "total")
#                                                             ) + ggtitle("Corrected log-normalized counts")

# # Compose figures using `patchwork`
# explanatory_variables_composed <- explanatory_variables_counts + explanatory_variables_logcounts_raw + explanatory_variables_logcounts + explanatory_variables_corrected +
#   plot_annotation(title = "Explanatory variables after normalisation and batch correction", tag_levels = "A", 
#                   theme = theme(plot.title = element_text(size = 18), plot.tag = element_text(size = 14)))
# 
# # Plot composed figures
# explanatory_variables_composed
# 
# # Save composed figures
# ggsave(filename = str_c(date, "_explanatory_variables_composed.pdf"),
#        plot = explanatory_variables_composed,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 12, height = 9, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
```

