---
title: "Topographic, single-cell gene expression profiling of Periaqueductal Gray neurons"
subtitle: "Part VIII: data visualisation"
author:
  - name: "Oriol Pavon Arocas, Sarah F. Olesen, and Tiago Branco"
    affiliation: "Sainsbury Wellcome Centre for Neural Circuits and Behaviour, University College London, UK"
    email: "oriol.pavon.16@ucl.ac.uk"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook:
    highlight: pygments
    number_sections: FALSE
    theme: lumen
    toc: TRUE
    toc_float: TRUE

#output: rmdformats::readthedown:
  #highlight: pygments
---

***

This is a pipeline to analyse single-cell RNA sequencing data from Periaqueductal Gray neurons (1) isolated from acute midbrain slices of transgenic mice using visually guided aspiration via patch pipettes and (2) processed using SMART-seq2 (Picelli et al., Nature Protocols 2014).

This pipeline has been generated after attending the [EMBL-EBI RNA-Sequence Analysis Course](https://www.ebi.ac.uk/training/events/2019/rna-sequence-analysis) and [attending](https://training.csx.cam.ac.uk/bioinformatics/event/2823386) and following the online course on [Analysis of single cell RNA-seq data](https://github.com/hemberg-lab/scRNA.seq.course) by the [Hemberg Lab](https://www.sanger.ac.uk/science/groups/hemberg-group). Many other resources have been used, including the [Orchestrating Single-Cell Analysis with Bioconductor book](https://bioconductor.org/books/release/OSCA/) by Robert Amezquita, Aaron Lun, Stephanie Hicks, and Raphael Gottardo, the [simpleSingleCell workflow in Bioconductor](https://bioconductor.org/packages/3.9/workflows/html/simpleSingleCell.html) maintained by Aaron Lun, the [rnaseqGene workflow](https://bioconductor.org/packages/3.10/workflows/html/rnaseqGene.html) maintained by Michael Love, the [RNAseq123 workflow](https://bioconductor.org/packages/3.10/workflows/html/RNAseq123.html) maintained by Matthew Ritchie, and the [EGSEA123 workflow](https://bioconductor.org/packages/3.10/workflows/html/EGSEA123.html) maintained by Matthew Ritchie.

Other key resources include [Bioconductor](http://www.bioconductor.org/) (Huber et al., Nature Methods 2015), [scRNA-tools](https://www.scrna-tools.org/), `scater` (McCarthy et al., Bioinformatics 2017), `scran` (Lun et al., F1000Res 2016), `SC3` (Kiselev et al., Nature Methods 2017), `Seurat` (Butler et al., Nature Biotechnology 2018), `clusterExperiment` (Risso et al., PLOS Computational Biology 2018), `limma` (Ritchie et al., Nucleic Acids Research 2015), `DESeq2` (Love et al., Genome Biology 2014), `edgeR` (Robinson et al., Bioinformatics 2010), `MAST` (Finak, McDavid, Yajima et al., Genome Biology 2015), `iSEE` (Rue-Albrecht & Marini et al., F1000Research 2018), `t-SNE` (van der Maaten & Hinton, Journal of Machine Learning Research 2008), `UMAP` (McInnes et al., arXiv 2018), and the [Mathematical Statistics and Machine Learning for Life Sciences](https://towardsdatascience.com/tagged/stats-ml-life-sciences) column by Nikolay Oskolkov.

***

## Step 8.1 | UMAP projection with clustering results
```{r}
# Set the directory where your data and scripts are:
setwd("D:/Dropbox (UCL - SWC)/Project_transcriptomics/analysis/PAG_scRNAseq_analysis")

# Set the path to save figures from this Part:
path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part8_datavis/"
date <- Sys.Date()
date <- gsub("-", "_", date)

# Load packages:
library(tidyverse)
library(SingleCellExperiment)
library(scater)
library(scran)
library(ggplot2)
library(pheatmap)
library(DESeq2)
library(edgeR)
library(limma)
library(monocle)
library(MAST)
library(ROCR)
library(patchwork)
library(extrafont)
library(RColorBrewer)
library(viridis)
```

To recapitulate the plots with the clustering results, we load the `PAG_sceset_qc_norm_filt_corr_clust` after normalisation, filtering, batch correction, and clustering. We should thus have a `corrected` slot in `assays`:
```{r}
# If starting from stored results, load saved filtered dataset from previous Step:
options(stringsAsFactors = FALSE)

PAG_sceset_qc_norm_filt_corr_clust <- readRDS("PAG_sceset_qc_norm_filt_corr_clust.rds") # Contains filtered cells and genes, log-normalised, filtered, corrected, and clustered data
assayNames(PAG_sceset_qc_norm_filt_corr_clust)
reducedDimNames(PAG_sceset_qc_norm_filt_corr_clust)
```

To remind ourselves about which clustering results we have available, we can look at the names of the `colData` slot:
```{r}
names(colData(PAG_sceset_qc_norm_filt_corr_clust)[grep("cluster", names(colData(PAG_sceset_qc_norm_filt_corr_clust)))])
```

We can quickly replot the results of our chosen clustering solutions:
```{r}
# Set theme parameters
theme_args_clustering <- theme(plot.title = element_text(size = 11, face = "bold"),
                               axis.title = element_text(size = 11, face = "plain"),
                               axis.text = element_text(size = 10, face = "plain"), 
                               legend.text = element_text(size = 9, face = "plain"), 
                               strip.text = element_text(size = 10, face = "plain"))


### Using UMAP (CV2) - 15_neighbors_0.01_min_dist ###
UMAP_cv2_15_001_PAGsubdivisions <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                  dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                  colour_by = "PAG.area", shape_by = "cell.type",
                                                  point_alpha = 0.6, point_size = 2, theme_size = 11
                                                  ) + labs(title = "PAG subdivisions", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

### Graph-based clustering results ###
# Using HVG from CV2 with rank-based weights
UMAP_cv2_15_001_RankWalktrap_k7 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                  dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                  colour_by = "SNN_clusters_cv2_rank_k7", shape_by = "cell.type", text_by = "SNN_clusters_cv2_rank_k7",
                                                  point_alpha = 0.6, point_size = 2, theme_size = 11
                                                  ) + labs(title = "Rank-Walktrap (k = 7)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_RankWalktrap_k12 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                  dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                  colour_by = "SNN_clusters_cv2_rank_k12", shape_by = "cell.type", text_by = "SNN_clusters_cv2_rank_k12",
                                                  point_alpha = 0.6, point_size = 2, theme_size = 11
                                                  ) + labs(title = "Rank-Walktrap (k = 12)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

# Using HVG from CV2 with jaccard-based weights
UMAP_cv2_15_001_JaccardLouvain_k5 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k5", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k5",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 5)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_JaccardLouvain_k8 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k8", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k8",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 8)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_JaccardLouvain_k9 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k9", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k9",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 9)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

UMAP_cv2_15_001_JaccardLouvain_k13 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "SNN_clusters_cv2_jaccard_k13", shape_by = "cell.type", text_by = "SNN_clusters_cv2_jaccard_k13",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "Jaccard-Louvain (k = 13)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering

### K-means clustering results ###
### Hierarchical clustering results ###
# The clusters obtained with both the K-means and the hierarchical clustering approach do not match well with the UMAP representation of the data. We exclude them from here but can be seen in Part 6 and readded if necessary.

### SC3 clustering results ###
UMAP_cv2_15_001_SC3_4 <- plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
                                                    dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
                                                    colour_by = "sc3_4_clusters", shape_by = "cell.type", text_by = "sc3_4_clusters",
                                                    point_alpha = 0.6, point_size = 2, theme_size = 11
                                                    ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2") + theme_args_clustering



# Prepare and save composed plots
library(patchwork)

UMAP_SNN_RankWalktrap_k7 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_RankWalktrap_k7) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_RankWalktrap_k7

UMAP_SNN_RankWalktrap_k12 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_RankWalktrap_k12) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_RankWalktrap_k12

UMAP_SNN_JaccardLouvain_k5 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k5) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k5

UMAP_SNN_JaccardLouvain_k8 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k8) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k8

UMAP_SNN_JaccardLouvain_k9 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k9) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k9

UMAP_SNN_JaccardLouvain_k13 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_JaccardLouvain_k13) +
  plot_annotation(title = "Graph-based clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SNN_JaccardLouvain_k13

UMAP_SC3_k4 <- (UMAP_cv2_15_001_PAGsubdivisions + UMAP_cv2_15_001_SC3_4) +
  plot_annotation(title = "SC3 clustering on UMAP (CV2)", tag_levels = "A", 
                  theme = theme(plot.title = element_text(size = 12, face = "bold"), plot.tag = element_text(size = 11))) + 
  plot_layout(ncol = 2, guides = "collect")
UMAP_SC3_k4


# # We could save the plots as follows, although we have already done so in earlier parts of the pipeline.
# ggsave(filename = str_c(date, "_UMAP_SNN_rank_k7_clusters.pdf"),
#        plot = UMAP_SNN_RankWalktrap_k7,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_rank_k12_clusters.pdf"),
#        plot = UMAP_SNN_RankWalktrap_k12,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k5_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k5,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k8_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k8,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k9_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k9,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_SNN_jaccard_k13_clusters.pdf"),
#        plot = UMAP_SNN_JaccardLouvain_k13,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
# 
# ggsave(filename = str_c(date, "_UMAP_sc3_k4_clusters.pdf"),
#        plot = UMAP_SC3_k4,
#        device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
#        path = path_for_figures,
#        width = 8, height = 4, units = "in", dpi = 300,
#        family = "Arial", bg = "transparent"
#        )
```

## Step 8.2 | Visualising results from differential expression analysis
```{r}
# Set the directory where your data and scripts are:
setwd("D:/Dropbox (UCL - SWC)/Project_transcriptomics/analysis/PAG_scRNAseq_analysis")

# Set the path to save figures from this Part:
path_for_figures <- "D:/Dropbox (UCL - SWC)/Project_transcriptomics/figures/R_figures_Part8_datavis/"
date <- Sys.Date()
date <- gsub("-", "_", date)

# Load packages:
library(tidyverse)
library(SingleCellExperiment)
library(scater)
library(scran)
library(ggplot2)
library(pheatmap)
library(DESeq2)
library(edgeR)
library(limma)
library(monocle)
library(MAST)
library(ROCR)
library(patchwork)
library(extrafont)
```

To visualise the results from differential expression analysis and marker genes, we load the `PAG_sceset_qc_norm_filt_corr_clust_de` after normalisation, filtering, batch correction, clustering, and DE analysis. We should thus have a `corrected` slot in `assays` and have removed the cells falling into the macrophage-enriched cluster:
```{r}
# If starting from stored results, load saved filtered dataset from previous Step:
options(stringsAsFactors = FALSE)

PAG_sceset_qc_norm_filt_corr_clust_de <- readRDS("PAG_sceset_qc_norm_filt_corr_clust_de.rds") # Contains filtered cells and genes, log-normalised, filtered, corrected, clustered data, and marker genes
assayNames(PAG_sceset_qc_norm_filt_corr_clust_de)
reducedDimNames(PAG_sceset_qc_norm_filt_corr_clust_de)
```

We can take a look at the lists of genes we have stored in the `metadata`, including the outputs from the different `findMarkers()` approaches:
```{r}
names(metadata(PAG_sceset_qc_norm_filt_corr_clust_de))
```

We can define our color palettes as follows:
```{r}
# Colours PAG
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

colours_celltype <- c(`VGAT` = "#FF8080", `VGluT2` = "#0F99B2")

colours_celltype_PAGarea <- c(`dmpag_VGAT` = "red", `dlpag_VGAT` = "coral2", `lpag_VGAT` = "salmon", `vlpag_VGAT` = "darksalmon",
                              `dmpag_VGluT2` = "cornflowerblue", `dlpag_VGluT2` = "cyan3", `lpag_VGluT2` = "aquamarine3", `vlpag_VGluT2` = "cadetblue3")

colours_PAGarea <- c(`dmPAG` = "#729ECE", `dlPAG` = "#FF9E4A", `lPAG` = "#67BF5C", `vlPAG` = "#ED665D")

# Heatmap palettes
heatmap_viridis <- rev(viridis::magma(100))
heatmap_blue <- colorRampPalette(c("white", "blue"))(100)
heatmap_VGAT <- colorRampPalette(c("white", "#FF8080"))(100)
heatmap_VGluT2 <- colorRampPalette(c("white", "#0F99B2"))(100)

# Clustering colours:
cluster_colours_10 <- c(`1` = "#729ECE", `2` = "#FF9E4A", `3` = "#67BF5C", `4` = "#ED665D", `5` = "#AD8BC9",
                        `6` = "#A8786E", `7` = "#ED97CA", `8` = "#A2A2A2", `9` = "#CDCC5D", `10` = "#6DCCDA")

cluster_colours_20 <- c(`1` = "#1F77B4", `2` = "#AEC7E8", `3` = "#FF7F0E", `4` = "#FFBB78", `5` = "#2CA02C",
                        `6` = "#98DF8A", `7` = "#D62728", `8` = "#FF9896", `9` = "#9467BD", `10` = "#C5B0D5",
                        `11` = "#8C564B", `12` = "#C49C94", `13` = "#E377C2", `14` = "#F7B6D2", `15` = "#7F7F7F", 
                        `16` = "#C7C7C7", `17` = "#BCBD22", `18` = "#DBDB8D", `19` = "#17BECF", `20` = "#9EDAE5")

cluster_colours <- scater:::.get_palette("tableau10medium") # hidden scater colours


# Scater palettes:
tableau20 <- c("#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78", "#2CA02C",
              "#98DF8A", "#D62728", "#FF9896", "#9467BD", "#C5B0D5",
              "#8C564B", "#C49C94", "#E377C2", "#F7B6D2", "#7F7F7F",
              "#C7C7C7", "#BCBD22", "#DBDB8D", "#17BECF", "#9EDAE5")

tableau10medium <- c("#729ECE", "#FF9E4A", "#67BF5C", "#ED665D",
                    "#AD8BC9", "#A8786E", "#ED97CA", "#A2A2A2",
                    "#CDCC5D", "#6DCCDA")

colorblind10 <- c("#006BA4", "#FF800E", "#ABABAB", "#595959",
                 "#5F9ED1", "#C85200", "#898989", "#A2C8EC",
                 "#FFBC79", "#CFCFCF")

trafficlight <- c("#B10318", "#DBA13A", "#309343", "#D82526",
                 "#FFC156", "#69B764", "#F26C64", "#FFDD71",
                 "#9FCD99")

purplegray12 <- c("#7B66D2", "#A699E8", "#DC5FBD", "#FFC0DA",
                 "#5F5A41", "#B4B19B", "#995688", "#D898BA",
                 "#AB6AD5", "#D098EE", "#8B7C6E", "#DBD4C5")

bluered12 <- c("#2C69B0", "#B5C8E2", "#F02720", "#FFB6B0", "#AC613C",
              "#E9C39B", "#6BA3D6", "#B5DFFD", "#AC8763", "#DDC9B4",
              "#BD0A36", "#F4737A")

greenorange12 <- c("#32A251", "#ACD98D", "#FF7F0F", "#FFB977",
                  "#3CB7CC", "#98D9E4", "#B85A0D", "#FFD94A",
                  "#39737C", "#86B4A9", "#82853B", "#CCC94D")

cyclic <-c("#1F83B4", "#1696AC", "#18A188", "#29A03C", "#54A338",
           "#82A93F", "#ADB828", "#D8BD35", "#FFBD4C", "#FFB022",
           "#FF9C0E", "#FF810E", "#E75727", "#D23E4E", "#C94D8C",
           "#C04AA7", "#B446B3", "#9658B1", "#8061B4", "#6F63BB")
```

### 8.2.0 | Exploring marker genes
We can first explore the differentially expressed genes identified by `findMarkers` by using the relevant `metadata` slot. However, these are all the gene names that passed the FDR threshold set in the previous part, and we have no other information other than the fact that they are sorted by p value.
```{r}
# Take a look at the gene lists we have in the metadata
names(metadata(PAG_sceset_qc_norm_filt_corr_clust_de))
```

An alternative way to explore marker genes is to load the .RDS file containing the results. This object has a slot for each comparison we carried out and will include the gene names together with their `p.value`, `FDR`, and `AUC`. We can use this to further filter and select them.
```{r}
cell_type_marker_genes_wilcox_all_up <- readRDS("D:\\Dropbox (UCL)\\Project_transcriptomics\\figures\\R_figures_Part7_DE\\cell_type\\cell_type_marker_genes_wilcox_all_up.rds")
names(cell_type_marker_genes_wilcox_all_up) # slots in the object
names(cell_type_marker_genes_wilcox_all_up$VGAT) # results in a specific slot
head(cell_type_marker_genes_wilcox_all_up$VGAT) # top markers for VGAT cells
head(cell_type_marker_genes_wilcox_all_up$VGluT2) # top markers for VGluT2 cells
```

As a reminder, the Wilcoxon rank sum test directly assesses separation between the expression distributions of different clusters and its test statistic is proportional to the area-under-the-curve (AUC), i.e., the concordance probability, which is the probability of a random cell from one cluster having higher expression than a random cell from another cluster. In a pairwise comparison, AUCs of 1 or 0 indicate that the two clusters have perfectly separated expression distributions. A value greater than 0.5 indicates that the gene is upregulated in the current cluster compared to the other cluster, while values less than 0.5 correspond to downregulation. We would typically expect AUCs of 0.7-0.8 for a strongly upregulated candidate marker. We can thus use the `AUC` value together with the `FDR` to select a subset of genes:
```{r}
# Filter by AUC and FDR threshold
AUC_marker_threshold <- 0.8
fdr_marker_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

cat("--- VGAT marker genes ---\n")
VGAT_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_marker_threshold), ])
VGAT_marker_genes
cat("\n--- VGluT2 marker genes ---\n")
VGluT2_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_marker_threshold), ])
VGluT2_marker_genes
```

We can then use the relevant gene list in our metadata slots to further subclassify and subset the interesting genes we want to plot:
```{r}
# Filter by AUC and FDR threshold
AUC_interesting_threshold <- 0.6
fdr_interesting_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

VGAT_de_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_interesting_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_interesting_threshold), ])

VGluT2_de_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_interesting_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_interesting_threshold), ])

# Print gene names from different families
cat("--- Ion Channels ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.ionchannels]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.ionchannels]

cat("\n\n--- Potassium Channels ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icPotassium]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icPotassium]

cat("\n\n--- Calcium Channels ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icCalcium]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icCalcium]

cat("\n\n--- Neuromodulators and Neuropeptides ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.NeuromodulatorsPeptides]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.NeuromodulatorsPeptides]

cat("\n\n--- Monoamines ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Monoamines]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Monoamines]

cat("\n\n--- Neurotransmitters ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Neurotransmitters]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Neurotransmitters]

cat("\n\n--- Transporters ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transporters]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transporters]

cat("\n\n--- Transcription Factors ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transcriptionfactors]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transcriptionfactors]
```

### 8.2.1 | Heatmaps 
One way to visualise the results from differentially expressed genes is via a heatmap. We will first take a look at the backbone code to generate a heatmap:
```{r}
library(pheatmap)
library(RColorBrewer)
library(viridis)

# Define plot title
plot_title <- "Marker genes for cell-type"

# Define thresholds and list of genes to plot
AUC_marker_threshold <- 0.8
fdr_marker_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

VGAT_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_marker_threshold), ])

VGluT2_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_marker_threshold), ])

# Plot heatmap
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(VGAT_marker_genes, VGluT2_marker_genes), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10
            )

# Plot heatmap centering expression values
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(VGAT_marker_genes, VGluT2_marker_genes), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(-5, 5), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10
            )
```

We can also make two heatmaps and colour each with a different color palette reflecting the cell type. We could use them to generate a plot in which half of the heatmaps is coloured according to one cell type and the other half is coloured according to the other:
```{r}
library(pheatmap)
library(RColorBrewer)
library(viridis)

# Define plot title and colours
plot_title <- "Marker genes for cell-type"
heatmap_VGAT <- colorRampPalette(c("white", "#FF8080"))(100)
heatmap_VGluT2 <- colorRampPalette(c("white", "#0F99B2"))(100)

# Define thresholds and list of genes to plot
AUC_marker_threshold <- 0.8
fdr_marker_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

VGAT_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_marker_threshold), ])

VGluT2_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_marker_threshold), ])

# Plot heatmap with VGAT colours
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(VGAT_marker_genes, VGluT2_marker_genes), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            color = heatmap_VGAT, # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10
            )

# Plot heatmap with VGluT2 colours
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(VGAT_marker_genes, VGluT2_marker_genes), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            color = heatmap_VGluT2, # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10, family = "Arial", bg = "transparent"
            )
```

Finally, for more control, we can also make a heatmap in which we customise the colors of the annotations (from https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/SingleCell/practical/crukBioinfoSummerSchoolJuly2018_scRnaSeqCellPopId_practical_run.html#192_detecting_genes_differentially_expressed_between_clusters):
```{r}
library(pheatmap)
library(RColorBrewer)
library(viridis)

# Define plot title
plot_title <- "Marker genes for cell-type"

# Define thresholds and list of genes to plot
AUC_marker_threshold <- 0.8
fdr_marker_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

VGAT_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_marker_threshold), ])

VGluT2_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_marker_threshold), ])

# Make a copy of the data containing only the selected genes
tmpData <- logcounts(PAG_sceset_qc_norm_filt_corr_clust_de)[c(VGAT_marker_genes, VGluT2_marker_genes),]
#dim(tmpData)

# Make dataframe with annotations for columns (cells) and cell ID:
mat_col <- data.frame(cell.type = PAG_sceset_qc_norm_filt_corr_clust_de$cell.type,
                      PAG.area = PAG_sceset_qc_norm_filt_corr_clust_de$PAG.area,
                      cell.type_PAG.subdivision = PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea#,
                      #cluster.ID = PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters
                      ) # annotations
rownames(mat_col) <- colnames(PAG_sceset_qc_norm_filt_corr_clust_de) # cell IUD
#dim(mat_col)

# Define colours
colours_celltype <- c(`VGAT` = "#FF8080", `VGluT2` = "#0F99B2")
colours_PAGarea <- c(`dmpag` = "#729ECE", `dlpag` = "#FF9E4A", `lpag` = "#67BF5C", `vlpag` = "#ED665D")
colours_celltype_PAGarea <- tableau10medium[1:length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))]
#colourCount_clusters = length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters)) # choose relevant clustering result

mat_colors <- list(cell.type = colours_celltype,
                   PAG.area = colours_PAGarea, # alternatively do tableau10medium[1:length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$PAG.area))]
                   cell.type_PAG.subdivision = colours_celltype_PAGarea#, 
                   #cluster.ID = tableau10medium[1:length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters))] # choose relevant clustering result
                   )

# If names haven't been defined (e.g. when using tableau10medium[1:length(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters))])
#names(mat_colors$cell.type) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))
#names(mat_colors$PAG.subdivision) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$PAG.area))
names(mat_colors$cell.type_PAG.subdivision) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))
#names(mat_colors$cluster.ID) <- sort(unique(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters))
#mat_colors

# Plot heatmap
pheatmap(tmpData[,order(mat_col$cell.type_PAG.subdivision)], # choose how to order the columns
         border_color = NA,
         scale = "none",
         #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
         show_rownames = TRUE, show_colnames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE,
         annotation_col = mat_col[,1:2], 
         annotation_colors = mat_colors[1:2],
         main = plot_title
         )
```

#### Marker genes for cell-type
We are now ready to generate a heatmap with a list of genes from our differential expression analysis. We will start by choosing from those genes that were strongly upregulated in either excitatory or inhibitory cells:
```{r}
library(pheatmap)
library(RColorBrewer)
library(viridis)

# Load data
cell_type_marker_genes_wilcox_all_up <- readRDS("D:\\Dropbox (UCL)\\Project_transcriptomics\\figures\\R_figures_Part7_DE\\cell_type\\cell_type_marker_genes_wilcox_all_up.rds")

# Define plot title
plot_title <- "Marker genes for cell-type"
save_name <- "heatmap_markers_celltype"

# Define thresholds and list of genes to plot
AUC_marker_threshold <- 0.75
fdr_marker_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

VGAT_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_marker_threshold), ])

VGluT2_marker_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_marker_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_marker_threshold), ])

# Plot heatmap
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(VGAT_marker_genes[1:6], VGAT_marker_genes[8], 
                         VGAT_marker_genes[10], VGluT2_marker_genes[1:8]), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10
            )

# Save heatmap
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(VGAT_marker_genes[1:6], VGAT_marker_genes[8], 
                         VGAT_marker_genes[10], VGluT2_marker_genes[1:8]), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10,
            width = 8, height = 4, family = "Arial", bg = "transparent",
            filename = paste(path_for_figures, "/heatmaps/", save_name, ".pdf", sep = "")
            )
```

#### Marker genes for clusters from SC3 (k = 4)
We can repeat the code above but this time focusing on genes from the clustering results obtained with `SC3`:
```{r}
library(pheatmap)
library(RColorBrewer)
library(viridis)

# Load data
SC3_clusters_k4_marker_genes_wilcox_all_up <- readRDS("D:\\Dropbox (UCL)\\Project_transcriptomics\\figures\\R_figures_Part7_DE\\SC3_clusters\\SC3_clusters_k4_marker_genes_wilcox_all_up.rds")

# Define plot title
plot_title <- "Marker genes for clusters from SC3 (k = 4)"
save_name <- "heatmap_markers_SC3_k4_clusters"

# Define thresholds and list of genes to plot
AUC_marker_threshold <- 0.75
fdr_marker_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

cluster_1_marker_genes <- rownames(SC3_clusters_k4_marker_genes_wilcox_all_up$`1`[(SC3_clusters_k4_marker_genes_wilcox_all_up$`1`$FDR<fdr_marker_threshold & 
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`1`$AUC.2>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`1`$AUC.3>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`1`$AUC.4>AUC_marker_threshold), ])
cluster_2_marker_genes <- rownames(SC3_clusters_k4_marker_genes_wilcox_all_up$`2`[(SC3_clusters_k4_marker_genes_wilcox_all_up$`2`$FDR<fdr_marker_threshold & 
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`2`$AUC.1>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`2`$AUC.3>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`2`$AUC.4>AUC_marker_threshold), ])
cluster_3_marker_genes <- rownames(SC3_clusters_k4_marker_genes_wilcox_all_up$`3`[(SC3_clusters_k4_marker_genes_wilcox_all_up$`3`$FDR<fdr_marker_threshold & 
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`3`$AUC.1>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`3`$AUC.2>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`3`$AUC.4>AUC_marker_threshold), ])
cluster_4_marker_genes <- rownames(SC3_clusters_k4_marker_genes_wilcox_all_up$`4`[(SC3_clusters_k4_marker_genes_wilcox_all_up$`4`$FDR<fdr_marker_threshold & 
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`4`$AUC.1>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`4`$AUC.2>AUC_marker_threshold &
                                                                                     SC3_clusters_k4_marker_genes_wilcox_all_up$`4`$AUC.3>AUC_marker_threshold), ])

# Plot heatmap
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(cluster_1_marker_genes, cluster_2_marker_genes[1:2], cluster_3_marker_genes[1:4], 
                         cluster_4_marker_genes[1:2], cluster_4_marker_genes[4]), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("sc3_4_clusters", "cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10
            )

# Save heatmap
plotHeatmap(PAG_sceset_qc_norm_filt_corr_clust_de,
            features = c(cluster_1_marker_genes, cluster_2_marker_genes[1:2], cluster_3_marker_genes[1:4], 
                         cluster_4_marker_genes[1:2], cluster_4_marker_genes[4]), # vector of row names specifying rows of object to show in the heatmap.
            columns = order(PAG_sceset_qc_norm_filt_corr_clust_de$sc3_4_clusters), # subset of columns to show in heatmap (or column order if cluster_cols=FALSE and order_columns_by=NULL). Defaults to NULL.
            exprs_values = "logcounts", # assay of object to be used as expression values for colouring the heatmap.
            #center = TRUE, # whether each row should have its mean expression centered at zero prior to plotting. Useful for examining logFC from the average across all cells.
            zlim = c(0, 12), # upper and lower bounds for the expression values (after centering). Defaults to the range of the expression matrix. Preserves the dynamic range of colours in the presence of outliers. Defaults to NULL.
            #symmetric = TRUE, # If TRUE, make zlim symmetric around zero (compute the maximum absolute value of zlim and multipy by c(-1, 1) to redefine zlim). Default is FALSE.
            #color = colorRampPalette(viridis(n=10))(100), # defaults to heatmap color palette: colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
            colour_columns_by = c("sc3_4_clusters", "cell.type", "PAG.area"), # how the columns should be annotated with colours. Can add clustering "sc3_4_clusters"
            show_rownames = TRUE, show_colnames = FALSE,
            cluster_rows = FALSE, cluster_cols = FALSE,
            main = plot_title, fontsize = 10,
            width = 8, height = 4, family = "Arial", bg = "transparent",
            filename = paste(path_for_figures, "/heatmaps/", save_name, ".pdf", sep = "")
            )
```

We could use some of the marker genes to colour the UMAP plots stored in the `SingleCellExperiment` object:
```{r}
plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
               dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
               colour_by = "Sncg", shape_by = "cell.type",
               point_alpha = 0.6, point_size = 2, theme_size = 11
               ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2")

plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
               dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
               colour_by = "Sox14", shape_by = "cell.type",
               point_alpha = 0.6, point_size = 2, theme_size = 11
               ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2")

plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
               dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
               colour_by = "Pax5", shape_by = "cell.type",
               point_alpha = 0.6, point_size = 2, theme_size = 11
               ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2")

plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
               dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
               colour_by = "Tcf7l2", shape_by = "cell.type",
               point_alpha = 0.6, point_size = 2, theme_size = 11
               ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2")

plotReducedDim(PAG_sceset_qc_norm_filt_corr_clust, 
               dimred = "UMAP_cv2_corrected_15_neighbors_0.01_min_dist",
               colour_by = "Tac1", shape_by = "cell.type",
               point_alpha = 0.6, point_size = 2, theme_size = 11
               ) + labs(title = "SC3 (k = 4)", x = "UMAP 1", y = "UMAP 2")

```

### 8.2.2 | Dot Plots
Besides heatmaps, another way we can visualise specific marker genes or genes of interest is through dotPlots.
```{r}
# Take a look at the gene lists we have in the metadata
names(metadata(PAG_sceset_qc_norm_filt_corr_clust_de))[1:35]
```

We can first explore the different types of potentially interesting genes we find and select a subset for plotting:
```{r}
# Load data
cell_type_marker_genes_wilcox_all_up <- readRDS("D:\\Dropbox (UCL)\\Project_transcriptomics\\figures\\R_figures_Part7_DE\\cell_type\\cell_type_marker_genes_wilcox_all_up.rds")

# Filter by AUC and FDR threshold
AUC_interesting_threshold <- 0 # Default to 0 to ignore this filter
fdr_interesting_threshold <- 0.05 # try 0.1, 0.05, 0.01 depending on how stringent you want your results to be

VGAT_de_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGAT[(cell_type_marker_genes_wilcox_all_up$VGAT$FDR<fdr_interesting_threshold & cell_type_marker_genes_wilcox_all_up$VGAT$AUC.VGluT2>AUC_interesting_threshold), ])

VGluT2_de_genes <- rownames(cell_type_marker_genes_wilcox_all_up$VGluT2[(cell_type_marker_genes_wilcox_all_up$VGluT2$FDR<fdr_interesting_threshold & cell_type_marker_genes_wilcox_all_up$VGluT2$AUC.VGAT>AUC_interesting_threshold), ])

# Print gene names from different families
cat("--- Ion Channels ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.ionchannels]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.ionchannels]

cat("\n\n--- Potassium Channels ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icPotassium]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icPotassium]

cat("\n\n--- Non-potassium Channels ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icRest]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icRest]

cat("\n\n--- Neurotransmitters ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Neurotransmitters]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Neurotransmitters]

cat("\n\n--- Neuromodulators and Neuropeptides ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.NeuromodulatorsPeptides]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.NeuromodulatorsPeptides]

cat("\n\n--- Transporters ---\n")
cat("--- VGAT cells\n")
VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transporters]
cat("\n--- VGluT2 cells\n")
VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transporters]

# cat("\n\n--- Transcription Factors ---\n")
# cat("--- VGAT cells\n")
# VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transcriptionfactors]
# cat("\n--- VGluT2 cells\n")
# VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Transcriptionfactors]
```

Once we have examined the potential marker genes, we can assign them to variables so they can easily be used to generate plots:
```{r}
# Potassium channels
potassium_channels_genes <- c(VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icPotassium], 
                              VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icPotassium])

# Calcium channels
calcium_channels_genes <- c(VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icCalcium], 
                            VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icCalcium])

# Asic channels
asic_channels_genes <- metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icAsic[(metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.icAsic %in% rownames(PAG_sceset_qc_norm_filt_corr_clust_de))]

# Other ion channels
other_channels_genes <- c("Scn7a", "Trpv6", "Ryr1", "P2rx6", "Aqp6")

# Neurotransmitters
neurotransmitter_genes <- c(VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Neurotransmitters], 
                            VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.Neurotransmitters])

# Neuromodulators and neuropeptides
nm_np_genes <- c(VGAT_de_genes[VGAT_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.NeuromodulatorsPeptides], 
                 VGluT2_de_genes[VGluT2_de_genes %in% metadata(PAG_sceset_qc_norm_filt_corr_clust_de)$genes.NeuromodulatorsPeptides])
```

#### Neuropeptides and neuromodulators
We will first generate plots with a subset of genes that are more expressed in one of the cell types but have similar expression across subdivisions.
```{r}
# Set theme parameters
theme_args_dots <- theme(plot.title = element_text(size = 11, face = "bold"),
                         axis.title = element_text(size = 11, face = "plain"),
                         axis.text = element_text(size = 10, face = "plain"), 
                         legend.text = element_text(size = 9, face = "plain"), 
                         strip.text = element_text(size = 10, face = "plain"))

# Define colours
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

# Define list of genes and title for plots
nm_np_gene_list <- c("Nxph1", "Pnoc", "Adcyap1", "Nxph4", "Tac1") #  c("Pnoc", "Oprl1")
#c("Drd1", "Rxfp2", "Tacr1", Prokr2", "Cartpt", "Grp", "Adra2a", "Spx", "Npy1r", "Lepr")
plot_title <- "Neuropeptides and Neuromodulators I"

# Make new order
# If we wanted a customised way to order the dotplot, we could create one as follows:
# PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea_dotplots <- factor(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea,
#                                                                           levels = c("VGAT_dmpag", "VGluT2_dmpag",  
#                                                                                      "VGAT_dlpag",  "VGluT2_dlpag", 
#                                                                                      "VGAT_lpag", "VGluT2_lpag", 
#                                                                                      "VGAT_vlpag", "VGluT2_vlpag"))

# Compose plots
nm_np_dotplot_vgat <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                               features = nm_np_gene_list,
                               exprs_values = "logcounts", detection_limit = 0,
                               group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                               low_color = "white", high_color = colour_vgat,
                               max_ave = NULL, max_detected = NULL
                               ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

nm_np_dotplot_vglut2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                 features = nm_np_gene_list,
                                 exprs_values = "logcounts", detection_limit = 0,
                                 group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                 low_color = "white", high_color = colour_vglut2,
                                 max_ave = NULL, max_detected = NULL
                                 ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

nm_np_dotplot_vgat_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                            features = nm_np_gene_list,
                                            exprs_values = "logcounts", detection_limit = 0,
                                            group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                            low_color = "white", high_color = colour_vgat,
                                            max_ave = NULL, max_detected = NULL
                                            ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

nm_np_dotplot_vglut2_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                              features = nm_np_gene_list,
                                              exprs_values = "logcounts", detection_limit = 0,
                                              group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                              low_color = "white", high_color = colour_vglut2,
                                              max_ave = NULL, max_detected = NULL
                                              ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

# Plot them
nm_np_dotplot_vgat
nm_np_dotplot_vglut2
nm_np_dotplot_vgat_subdivisions
nm_np_dotplot_vglut2_subdivisions

# Save plots
ggsave(filename = str_c(date, "_dotplot_nm_np_1_vgat.pdf"),
       plot = nm_np_dotplot_vgat,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_nm_np_1_vglut2.pdf"),
       plot = nm_np_dotplot_vglut2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_nm_np_1_vgat_subdivisions.pdf"),
       plot = nm_np_dotplot_vgat_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_nm_np_1_vglut2_subdivisions.pdf"),
       plot = nm_np_dotplot_vglut2_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

Now we will generate plots with a subset of genes that show a preferential expression for one or two subdivisions:
```{r}
# Set theme parameters
theme_args_dots <- theme(plot.title = element_text(size = 11, face = "bold"),
                         axis.title = element_text(size = 11, face = "plain"),
                         axis.text = element_text(size = 10, face = "plain"), 
                         legend.text = element_text(size = 9, face = "plain"), 
                         strip.text = element_text(size = 10, face = "plain"))

# Define colours
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

# Define list of genes and title for plots
nm_np_gene_list_2 <- c("Drd1", "Rxfp2", "Rxfp3", "Tacr1", "Galr1", "Prokr2", "Cartpt", "Grp", "Spx", "Adra1a", "Adra2a")
plot_title <- "Neuropeptides and Neuromodulators II"

# Make new order
# If we wanted a customised way to order the dotplot, we could create one as follows:
# PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea_dotplots <- factor(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea,
#                                                                           levels = c("VGAT_dmpag", "VGluT2_dmpag",  
#                                                                                      "VGAT_dlpag",  "VGluT2_dlpag", 
#                                                                                      "VGAT_lpag", "VGluT2_lpag", 
#                                                                                      "VGAT_vlpag", "VGluT2_vlpag"))

# Compose plots
nm_np_dotplot_vgat_2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                 features = nm_np_gene_list_2,
                                 exprs_values = "logcounts", detection_limit = 0,
                                 group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                 low_color = "white", high_color = colour_vgat,
                                 max_ave = NULL, max_detected = NULL
                                 ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list_2)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

nm_np_dotplot_vglut2_2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                   features = nm_np_gene_list_2,
                                   exprs_values = "logcounts", detection_limit = 0,
                                   group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                   low_color = "white", high_color = colour_vglut2,
                                   max_ave = NULL, max_detected = NULL
                                   ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list_2)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

nm_np_dotplot_vgat_subdivisions_2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                              features = nm_np_gene_list_2,
                                              exprs_values = "logcounts", detection_limit = 0,
                                              group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                              low_color = "white", high_color = colour_vgat,
                                              max_ave = NULL, max_detected = NULL
                                              ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list_2)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

nm_np_dotplot_vglut2_subdivisions_2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                                features = nm_np_gene_list_2,
                                                exprs_values = "logcounts", detection_limit = 0,
                                                group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                                low_color = "white", high_color = colour_vglut2,
                                                max_ave = NULL, max_detected = NULL
                                                ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(nm_np_gene_list_2)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

# Plot them
nm_np_dotplot_vgat_2
nm_np_dotplot_vglut2_2
nm_np_dotplot_vgat_subdivisions_2
nm_np_dotplot_vglut2_subdivisions_2

# Save plots
ggsave(filename = str_c(date, "_dotplot_nm_np_2_vgat.pdf"),
       plot = nm_np_dotplot_vgat_2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_nm_np_2_vglut2.pdf"),
       plot = nm_np_dotplot_vglut2_2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_nm_np_2_vgat_subdivisions.pdf"),
       plot = nm_np_dotplot_vgat_subdivisions_2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_nm_np_2_vglut2_subdivisions.pdf"),
       plot = nm_np_dotplot_vglut2_subdivisions_2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```
#### Dopamine receptors
```{r}
# Set theme parameters
theme_args_dots <- theme(plot.title = element_text(size = 11, face = "bold"),
                         axis.title = element_text(size = 11, face = "plain"),
                         axis.text = element_text(size = 10, face = "plain"), 
                         legend.text = element_text(size = 9, face = "plain"), 
                         strip.text = element_text(size = 10, face = "plain"))

# Define colours
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

# Define list of genes and title for plots
dopamine_gene_list <- c("Drd1", "Drd2", "Drd3", "Drd4", "Drd5")
plot_title <- "Dopamine receptors"

# Make new order
# If we wanted a customised way to order the dotplot, we could create one as follows:
# PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea_dotplots <- factor(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea,
#                                                                           levels = c("VGAT_dmpag", "VGluT2_dmpag",  
#                                                                                      "VGAT_dlpag",  "VGluT2_dlpag", 
#                                                                                      "VGAT_lpag", "VGluT2_lpag", 
#                                                                                      "VGAT_vlpag", "VGluT2_vlpag"))

# Compose plots
DA_dotplot_vgat <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                            features = dopamine_gene_list,
                            exprs_values = "logcounts", detection_limit = 0,
                            group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                            low_color = "white", high_color = colour_vgat,
                            max_ave = NULL, max_detected = NULL
                            ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(dopamine_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

DA_dotplot_vglut2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                              features = dopamine_gene_list,
                              exprs_values = "logcounts", detection_limit = 0,
                              group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                              low_color = "white", high_color = colour_vglut2,
                              max_ave = NULL, max_detected = NULL
                              ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(dopamine_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

DA_dotplot_vgat_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                         features = dopamine_gene_list,
                                         exprs_values = "logcounts", detection_limit = 0,
                                         group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                         low_color = "white", high_color = colour_vgat,
                                         max_ave = NULL, max_detected = NULL
                                         ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(dopamine_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

DA_dotplot_vglut2_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                           features = dopamine_gene_list,
                                           exprs_values = "logcounts", detection_limit = 0,
                                           group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                           low_color = "white", high_color = colour_vglut2,
                                           max_ave = NULL, max_detected = NULL
                                           ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(dopamine_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

# Plot them
DA_dotplot_vgat
DA_dotplot_vglut2
DA_dotplot_vgat_subdivisions
DA_dotplot_vglut2_subdivisions

# Save plots
ggsave(filename = str_c(date, "_dotplot_dopamine_vgat.pdf"),
       plot = DA_dotplot_vgat,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_dopamine_vglut2.pdf"),
       plot = DA_dotplot_vglut2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_dopamine_vgat_subdivisions.pdf"),
       plot = DA_dotplot_vgat_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_dopamine_vglut2_subdivisions.pdf"),
       plot = DA_dotplot_vglut2_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

#### Asic channels
```{r}
# Set theme parameters
theme_args_dots <- theme(plot.title = element_text(size = 11, face = "bold"),
                         axis.title = element_text(size = 11, face = "plain"),
                         axis.text = element_text(size = 10, face = "plain"), 
                         legend.text = element_text(size = 9, face = "plain"), 
                         strip.text = element_text(size = 10, face = "plain"))

# Define colours
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

# Define list of genes and title for plots
asic_gene_list <- c("Asic1", "Asic2", "Asic4")
plot_title <- "ASIC channels"

# Make new order
# If we wanted a customised way to order the dotplot, we could create one as follows:
# PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea_dotplots <- factor(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea,
#                                                                           levels = c("VGAT_dmpag", "VGluT2_dmpag",  
#                                                                                      "VGAT_dlpag",  "VGluT2_dlpag", 
#                                                                                      "VGAT_lpag", "VGluT2_lpag", 
#                                                                                      "VGAT_vlpag", "VGluT2_vlpag"))

# Compose plots
asic_dotplot_vgat <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                              features = asic_gene_list,
                              exprs_values = "logcounts", detection_limit = 0,
                              group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                              low_color = "white", high_color = colour_vgat,
                              max_ave = NULL, max_detected = NULL
                              ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(asic_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

asic_dotplot_vglut2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                features = asic_gene_list,
                                exprs_values = "logcounts", detection_limit = 0,
                                group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                low_color = "white", high_color = colour_vglut2,
                                max_ave = NULL, max_detected = NULL
                                ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(asic_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

asic_dotplot_vgat_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                           features = asic_gene_list,
                                           exprs_values = "logcounts", detection_limit = 0,
                                           group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                           low_color = "white", high_color = colour_vgat,
                                           max_ave = NULL, max_detected = NULL
                                           ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(asic_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

asic_dotplot_vglut2_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                             features = asic_gene_list,
                                             exprs_values = "logcounts", detection_limit = 0,
                                             group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                             low_color = "white", high_color = colour_vglut2,
                                             max_ave = NULL, max_detected = NULL
                                             ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(asic_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

# Plot them
asic_dotplot_vgat
asic_dotplot_vglut2
asic_dotplot_vgat_subdivisions
asic_dotplot_vglut2_subdivisions

# Save plots
ggsave(filename = str_c(date, "_dotplot_asic_vgat.pdf"),
       plot = asic_dotplot_vgat,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_asic_vglut2.pdf"),
       plot = asic_dotplot_vglut2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_asic_vgat_subdivisions.pdf"),
       plot = asic_dotplot_vgat_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_asic_vglut2_subdivisions.pdf"),
       plot = asic_dotplot_vglut2_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

#### Calcium channels
```{r}
# Set theme parameters
theme_args_dots <- theme(plot.title = element_text(size = 11, face = "bold"),
                         axis.title = element_text(size = 11, face = "plain"),
                         axis.text = element_text(size = 10, face = "plain"), 
                         legend.text = element_text(size = 9, face = "plain"), 
                         strip.text = element_text(size = 10, face = "plain"))

# Define colours
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

# Define list of genes and title for plots
calcium_gene_list <- c("Cacna1a", "Cacna1b", "Cacna1c", "Cacna1d", "Cacna1e", # alpha subunits
                       "Cacna2d1", "Cacna2d2") # a2d subunits
plot_title <- "Calcium channels"

# Make new order
# If we wanted a customised way to order the dotplot, we could create one as follows:
# PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea_dotplots <- factor(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea,
#                                                                           levels = c("VGAT_dmpag", "VGluT2_dmpag",  
#                                                                                      "VGAT_dlpag",  "VGluT2_dlpag", 
#                                                                                      "VGAT_lpag", "VGluT2_lpag", 
#                                                                                      "VGAT_vlpag", "VGluT2_vlpag"))

# Compose plots
calcium_dotplot_vgat <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                 features = calcium_gene_list,
                                 exprs_values = "logcounts", detection_limit = 0,
                                 group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                 low_color = "white", high_color = colour_vgat,
                                 max_ave = NULL, max_detected = NULL
                                 ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(calcium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

calcium_dotplot_vglut2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                   features = calcium_gene_list,
                                   exprs_values = "logcounts", detection_limit = 0,
                                   group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                   low_color = "white", high_color = colour_vglut2,
                                   max_ave = NULL, max_detected = NULL
                                   ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(calcium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

calcium_dotplot_vgat_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                              features = calcium_gene_list,
                                              exprs_values = "logcounts", detection_limit = 0,
                                              group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                              low_color = "white", high_color = colour_vgat,
                                              max_ave = NULL, max_detected = NULL
                                              ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(calcium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

calcium_dotplot_vglut2_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                                features = calcium_gene_list,
                                                exprs_values = "logcounts", detection_limit = 0,
                                                group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                                low_color = "white", high_color = colour_vglut2,
                                                max_ave = NULL, max_detected = NULL
                                                ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(calcium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

# Plot them
calcium_dotplot_vgat
calcium_dotplot_vglut2
calcium_dotplot_vgat_subdivisions
calcium_dotplot_vglut2_subdivisions

# Save plots
ggsave(filename = str_c(date, "_dotplot_calcium_vgat.pdf"),
       plot = calcium_dotplot_vgat,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_calcium_vglut2.pdf"),
       plot = calcium_dotplot_vglut2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_calcium_vgat_subdivisions.pdf"),
       plot = calcium_dotplot_vgat_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_calcium_vglut2_subdivisions.pdf"),
       plot = calcium_dotplot_vglut2_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```

#### Potassium channels
```{r}
# Set theme parameters
theme_args_dots <- theme(plot.title = element_text(size = 11, face = "bold"),
                         axis.title = element_text(size = 11, face = "plain"),
                         axis.text = element_text(size = 10, face = "plain"), 
                         legend.text = element_text(size = 9, face = "plain"), 
                         strip.text = element_text(size = 10, face = "plain"))

# Define colours
colour_vgat <- "#FF8080"
colour_vglut2 <-"#0F99B2"

# Define list of genes and title for plots
potassium_gene_list <- c("Kcnh1", # non-inactivating delayed rectifier (ether-a-go-go)
                         "Kcnh5", # non-inactivating outward rectifier (ether-a-go-go)
                         "Kcnq5", # M-type current, slowly activating and deactivating potassium conductance
                         "Kcnj3", # inward rectifier (GIRK-1)
                         "Kcnk9"  # pH-dependent, voltage-insensitive, background two-pore potassium channel (TASK3)
                         )
plot_title <- "potassium channels"

# Make new order
# If we wanted a customised way to order the dotplot, we could create one as follows:
# PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea_dotplots <- factor(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea,
#                                                                           levels = c("VGAT_dmpag", "VGluT2_dmpag",  
#                                                                                      "VGAT_dlpag",  "VGluT2_dlpag", 
#                                                                                      "VGAT_lpag", "VGluT2_lpag", 
#                                                                                      "VGAT_vlpag", "VGluT2_vlpag"))

# Compose plots
potassium_dotplot_vgat <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                   features = potassium_gene_list,
                                   exprs_values = "logcounts", detection_limit = 0,
                                   group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                   low_color = "white", high_color = colour_vgat,
                                   max_ave = NULL, max_detected = NULL
                                   ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(potassium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

potassium_dotplot_vglut2 <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                     features = potassium_gene_list,
                                     exprs_values = "logcounts", detection_limit = 0,
                                     group = "cell.type", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                     low_color = "white", high_color = colour_vglut2,
                                     max_ave = NULL, max_detected = NULL
                                     ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(potassium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$cell.type))

potassium_dotplot_vgat_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                                features = potassium_gene_list,
                                                exprs_values = "logcounts", detection_limit = 0,
                                                group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                                low_color = "white", high_color = colour_vgat,
                                                max_ave = NULL, max_detected = NULL
                                                ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(potassium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

potassium_dotplot_vglut2_subdivisions <- plotDots(PAG_sceset_qc_norm_filt_corr_clust_de,
                                                  features = potassium_gene_list,
                                                  exprs_values = "logcounts", detection_limit = 0,
                                                  group = "celltype_PAGarea", #"cell.type", "celltype_PAGarea", "celltype_PAGAPaxis"
                                                  low_color = "white", high_color = colour_vglut2,
                                                  max_ave = NULL, max_detected = NULL
                                                  ) + 
  labs(title = plot_title, x = NULL, y = NULL) + theme_args_dots +
  scale_y_discrete(limits = rev(potassium_gene_list)) + 
  scale_x_discrete(limits = levels(PAG_sceset_qc_norm_filt_corr_clust_de$celltype_PAGarea))

# Plot them
potassium_dotplot_vgat
potassium_dotplot_vglut2
potassium_dotplot_vgat_subdivisions
potassium_dotplot_vglut2_subdivisions

# Save plots
ggsave(filename = str_c(date, "_dotplot_potassium_vgat.pdf"),
       plot = potassium_dotplot_vgat,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_potassium_vglut2.pdf"),
       plot = potassium_dotplot_vglut2,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 4, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_potassium_vgat_subdivisions.pdf"),
       plot = potassium_dotplot_vgat_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
ggsave(filename = str_c(date, "_dotplot_potassium_vglut2_subdivisions.pdf"),
       plot = potassium_dotplot_vglut2_subdivisions,
       device = "pdf", # or one of "eps", "ps", "tex" (pictex), "pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).
       path = str_c(path_for_figures, "/dotplots"),
       width = 8, height = 4, units = "in", dpi = 300,
       family = "Arial", bg = "transparent"
       )
```